{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downloads Publication Information for PANGO Lineages from the CORD-19 Data Set\n",
    "**[Work in progress]**\n",
    "\n",
    "This notebook text-mines [PANGO lineage](https://cov-lineages.org/) mentions in the titles and abstracts of publications and preprints from the CORD-19 data set. Note, the text-mined results may contain false positive!\n",
    "\n",
    "Data sources: [PANGO Lineage Designations](https://github.com/cov-lineages/pango-designation), \n",
    "[CORD-19](https://allenai.org/data/cord-19)\n",
    "\n",
    "References:\n",
    "\n",
    "Rambaut A, et al., A dynamic nomenclature proposal for SARS-CoV-2 lineages to assist genomic epidemiology(2020) Nature Microbiology [doi:10.1038/s41564-020-0770-5](https://doi.org/10.1038/s41564-020-0770-5).\n",
    "\n",
    "Lucy Lu Wang, et al., CORD-19: The COVID-19 Open Research Dataset (2020) [arXiv:2004.10706v4](https://arxiv.org/abs/2004.10706).\n",
    "\n",
    "Author: Peter Rose (pwrose@ucsd.edu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import io\n",
    "import dateutil\n",
    "import re\n",
    "from pathlib import Path\n",
    "import nltk\n",
    "import json, requests\n",
    "from urllib.request import urlopen\n",
    "from xml.etree.ElementTree import parse\n",
    "import urllib\n",
    "import time\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/lyt/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-a1516f46-b63a-46dd-b67a-1fb59d6c5d05/import\n"
     ]
    }
   ],
   "source": [
    "NEO4J_IMPORT = \"/Users/lyt/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-a1516f46-b63a-46dd-b67a-1fb59d6c5d05/import\"#Path(os.getenv('NEO4J_IMPORT'))\n",
    "print(NEO4J_IMPORT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Publication-MENTIONS-Lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pd.read_csv('/Users/lyt/Desktop/covid19-kg2/data/Relationships/Publication-MENTIONS-Lineage.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>Q.3</td>\n",
       "      <td>Among the 732 clinical samples that were teste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>By WGS, these samples were variant of interest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>This strain clade 201/501Y.V1, Pango lineage B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>26\\n There are three mutations of significant ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>P.1</td>\n",
       "      <td>56\\n Safety and immunogenicity study of a SARS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Gamma variant is particularly found from the o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>These resembling variants, B.1.427 and B.1.429...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>These resembling variants, B.1.427 and B.1.429...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>B.1.617.2 (Delta variant) is a variant of line...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>83\\n The B.1.617.1 strain is 6.8‐fold more res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>92\\n Researchers discovered that this new C.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>K417N, N440K, G446S, S477N, T478K, E484A, Q493...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.625</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.348</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>C.37.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>C.37</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.487</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.621.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>P.1.10</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>P.1</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>The D614G substitution is found in most of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The D614G substitution is found in most of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>In addition, some of the notable substitutions...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>In addition, some of the notable substitutions...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>Fitness-enhancing mutations have been reported...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>However, in-vitro neutralization assays show t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>According to the CDC, US, the P.1 variant seem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Since the original strain of SARS-CoV-2 virus ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>P.1</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>*) and Omicron (B.1.1.529 and BA.*).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>Q.3</td>\n",
       "      <td>Among the 732 clinical samples that were teste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>By WGS, these samples were variant of interest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.519</td>\n",
       "      <td>*), and Omicron (B.1.1.519 and BA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>P.1</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>C.37</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>On 24 November 2021, the B.1.1.529 lineage, a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>P.1</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The second largest clade identified by Nextcla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.369</td>\n",
       "      <td>The second largest clade identified by Nextcla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>Pangolin classified most of these as the US st...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>A clear trend can be observed with the average...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.324</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1.317</td>\n",
       "      <td>The A845S spike mutation was observed at the s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>P.1</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>The most recent VOC, first identified in South...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.529</td>\n",
       "      <td>Their neutralization spectrum is different and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>Residues mutated in B.1.1.7, B.1.617 and B.1.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Although neutralization of one of these antibo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Although neutralization of one of these antibo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Finally, we investigate the effects of B.1.1.7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Similar histopathological findings of lung and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>C.37</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.214.2</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.36</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>P.1</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>C.37</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>R.1</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1</td>\n",
       "      <td>One of the earliest variants that is highly in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>Subsequent genomic surveillance has led to the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>The B.1.427/B.1.429 and B.1.617 lineage varian...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>P.3</td>\n",
       "      <td>The E484K substitution found in B.1.351, B.1.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>These findings extend a prior study reporting ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>Sequences corresponding to the S1/S2 region of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Genomic surveillance of SC2 variants has been ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>Here, we describe the introduction of SC2 line...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>A.2.4</td>\n",
       "      <td>In January 2021, BPHL detected the first cases...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The A.2.5 lineage in Florida became of particu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>A.2.5 also contains spike L452R, a CDC-defined...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>C.36</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Additional mutations causing increased transmi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The alpha/B.1.1.7 SARS-CoV-2 lineage emerged i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The incidence domination was likely due to a f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>The incidence domination was likely due to a f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Some of these residue substitutions, including...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Some of these residue substitutions, including...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>All the above VOC/VOI have been for the most p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1</td>\n",
       "      <td>In order to examine whether the increased affi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>P.1</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.1.298</td>\n",
       "      <td>Y453F increases the ACE2 binding affinity, and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.177</td>\n",
       "      <td>A222V and S477N are described in 20A.EU1 (B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.160</td>\n",
       "      <td>A222V and S477N are described in 20A.EU1 (B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.141</td>\n",
       "      <td>N439K is in B.1.141, B.1.258, and in mink stra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.258</td>\n",
       "      <td>N439K is in B.1.141, B.1.258, and in mink stra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.37</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.619</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.214.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.466.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.620</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.523</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.519</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>R.1</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.36.3</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Some former VOIs but no longer designated as V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Some former VOIs but no longer designated as V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>P.3 as well as P.1 belong to B.1.1.28 lineage ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>At the same time, B.1.525 was resistant to neu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>With the spread of the local epidemic in India...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The Delta (B.1.617.2) variant was first identi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>C.37</td>\n",
       "      <td>The Lambda variant belongs to the C.37 lineage...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>On August 30, 2021, the WHO announced the Mu v...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>PMC8501352</td>\n",
       "      <td>B.1.117</td>\n",
       "      <td>At the time of the first clinical trials the “...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>The massive emergence of L452R variants was fi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.232</td>\n",
       "      <td>By PCR amplification and Sanger sequencing of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.27</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.263</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.36</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.21</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.466</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.459</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.526.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.2.5.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.2.5.2</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.177.83</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.36.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>P.4</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.37</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.16</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.36.3</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>P.1</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>D.2</td>\n",
       "      <td>The D.2 variant in Australia was defined by an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.160</td>\n",
       "      <td>In contrast, B.1.160, which carries S S477N so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>The substitution D260Y, present in B.1.427 and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.177</td>\n",
       "      <td>Several lineages show introgression from other...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1</td>\n",
       "      <td>c. Population-level analysis of new mutations ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>V.2</td>\n",
       "      <td>Variants with minor allele frequency less than...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>C.14</td>\n",
       "      <td>We identified one previously reported (c.2729 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>V.1</td>\n",
       "      <td>All variants were reclassified according to AC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>C.38</td>\n",
       "      <td>Results: We found 8 mutations in this patient ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           from          to                                           evidence\n",
       "0    PMC9119066   B.1.1.529  Since the original strain of severe acute resp...\n",
       "1    PMC9119066   B.1.617.2  Since the original strain of severe acute resp...\n",
       "2    PMC9119066         P.1  Since the original strain of severe acute resp...\n",
       "3    PMC9119066     B.1.1.7  Since the original strain of severe acute resp...\n",
       "4    PMC9119066     B.1.351  Since the original strain of severe acute resp...\n",
       "5    PMC9119066         Q.3  Among the 732 clinical samples that were teste...\n",
       "6    PMC9119066     B.1.621  By WGS, these samples were variant of interest...\n",
       "7    PMC9088647     B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...\n",
       "8    PMC9088647     B.1.351  26\\n There are three mutations of significant ...\n",
       "9    PMC9088647         P.1  56\\n Safety and immunogenicity study of a SARS...\n",
       "10   PMC9088647         P.2  Gamma variant is particularly found from the o...\n",
       "11   PMC9088647     B.1.427  These resembling variants, B.1.427 and B.1.429...\n",
       "12   PMC9088647     B.1.429  These resembling variants, B.1.427 and B.1.429...\n",
       "13   PMC9088647     B.1.617  B.1.617.2 (Delta variant) is a variant of line...\n",
       "14   PMC9088647   B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...\n",
       "15   PMC9088647       C.1.2  92\\n Researchers discovered that this new C.1....\n",
       "16   PMC9088647   B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...\n",
       "17   PMC9008484     B.1.621  The three largest cities in Colombia: Bogotá, ...\n",
       "18   PMC9008484   B.1.617.2  The three largest cities in Colombia: Bogotá, ...\n",
       "19   PMC9008484         P.1  The three largest cities in Colombia: Bogotá, ...\n",
       "20   PMC9008484     B.1.1.7  The three largest cities in Colombia: Bogotá, ...\n",
       "21   PMC9008484       B.1.1  The figure shows the number of cases for SARS-...\n",
       "22   PMC9008484       A.2.5  The figure shows the number of cases for SARS-...\n",
       "23   PMC9008484     B.1.625  The figure shows the number of cases for SARS-...\n",
       "24   PMC9008484   B.1.1.348  The figure shows the number of cases for SARS-...\n",
       "25   PMC9008484      C.37.1  The figure shows the number of cases for SARS-...\n",
       "26   PMC9008484        C.37  The figure shows the number of cases for SARS-...\n",
       "27   PMC9008484   B.1.1.487  The figure shows the number of cases for SARS-...\n",
       "28   PMC9008484   B.1.621.1  The figure shows the number of cases for SARS-...\n",
       "29   PMC9008484      P.1.10  The figure shows the number of cases for SARS-...\n",
       "30   PMC9008484         B.1  The figure shows the number of cases for SARS-...\n",
       "31   PMC8902153   B.1.617.2  We also look at the effectiveness of approved ...\n",
       "32   PMC8902153         P.1  We also look at the effectiveness of approved ...\n",
       "33   PMC8902153     B.1.351  We also look at the effectiveness of approved ...\n",
       "34   PMC8902153   B.1.1.529  The D614G substitution is found in most of the...\n",
       "35   PMC8902153     B.1.1.7  The D614G substitution is found in most of the...\n",
       "36   PMC8902153     B.1.617  In addition, some of the notable substitutions...\n",
       "37   PMC8902153     B.1.429  In addition, some of the notable substitutions...\n",
       "38   PMC8902153       B.1.1  Fitness-enhancing mutations have been reported...\n",
       "39   PMC8902153   B.1.617.1  However, in-vitro neutralization assays show t...\n",
       "40   PMC8902153     B.1.427  According to the CDC, US, the P.1 variant seem...\n",
       "41    PPR452207     B.1.1.7  Since the original strain of SARS-CoV-2 virus ...\n",
       "42    PPR452207   B.1.617.2  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "43    PPR452207         P.1  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "44    PPR452207     B.1.351  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "45    PPR452207   B.1.1.529               *) and Omicron (B.1.1.529 and BA.*).\n",
       "46    PPR452207         Q.3  Among the 732 clinical samples that were teste...\n",
       "47    PPR452207     B.1.621  By WGS, these samples were variant of interest...\n",
       "48    PPR452207   B.1.1.519                 *), and Omicron (B.1.1.519 and BA.\n",
       "49   PMC8942852     B.1.621  As of December 2021, the World Health Organiza...\n",
       "50   PMC8942852   B.1.1.529  As of December 2021, the World Health Organiza...\n",
       "51   PMC8942852   B.1.617.2  As of December 2021, the World Health Organiza...\n",
       "52   PMC8942852         P.1  As of December 2021, the World Health Organiza...\n",
       "53   PMC8942852        C.37  As of December 2021, the World Health Organiza...\n",
       "54   PMC8942852     B.1.1.7  As of December 2021, the World Health Organiza...\n",
       "55   PMC8942852     B.1.351  As of December 2021, the World Health Organiza...\n",
       "56   PMC8942852       B.1.1  On 24 November 2021, the B.1.1.529 lineage, a ...\n",
       "57   PMC8769358     B.1.427  A total of 411 samples belonging to the Alpha ...\n",
       "58   PMC8769358   B.1.617.2  A total of 411 samples belonging to the Alpha ...\n",
       "59   PMC8769358         P.1  A total of 411 samples belonging to the Alpha ...\n",
       "60   PMC8769358     B.1.1.7  A total of 411 samples belonging to the Alpha ...\n",
       "61   PMC8769358     B.1.351  A total of 411 samples belonging to the Alpha ...\n",
       "62   PMC8769358         B.1  The second largest clade identified by Nextcla...\n",
       "63   PMC8769358     B.1.369  The second largest clade identified by Nextcla...\n",
       "64   PMC8769358       B.1.2  Pangolin classified most of these as the US st...\n",
       "65   PMC8769358     B.1.526  A clear trend can be observed with the average...\n",
       "66   PMC8769358       B.1.1  While this mutation was observed in the B.1.1....\n",
       "67   PMC8769358     B.1.621  While this mutation was observed in the B.1.1....\n",
       "68   PMC8769358     B.1.324  While this mutation was observed in the B.1.1....\n",
       "69   PMC8769358   B.1.1.317  The A845S spike mutation was observed at the s...\n",
       "70    PPR442789   B.1.617.2  As a result of viral evolution, the initial SA...\n",
       "71    PPR442789         P.1  As a result of viral evolution, the initial SA...\n",
       "72    PPR442789     B.1.1.7  As a result of viral evolution, the initial SA...\n",
       "73    PPR442789     B.1.351  As a result of viral evolution, the initial SA...\n",
       "74    PPR442789   B.1.1.529  The most recent VOC, first identified in South...\n",
       "75    PPR442789     B.1.529  Their neutralization spectrum is different and...\n",
       "76    PPR442789     B.1.617  Residues mutated in B.1.1.7, B.1.617 and B.1.3...\n",
       "77   PMC8690919     B.1.1.7  Although neutralization of one of these antibo...\n",
       "78   PMC8690919     B.1.351  Although neutralization of one of these antibo...\n",
       "79   PMC8690919   B.1.617.2  Finally, we investigate the effects of B.1.1.7...\n",
       "80   PMC8690919         P.1  Similar histopathological findings of lung and...\n",
       "81   PMC8754802         P.1  Furthermore, among these lineages, WHO identif...\n",
       "82   PMC8754802         P.3  Furthermore, among these lineages, WHO identif...\n",
       "83   PMC8754802     B.1.351  Furthermore, among these lineages, WHO identif...\n",
       "84   PMC8754802     B.1.621  Furthermore, among these lineages, WHO identif...\n",
       "85   PMC8754802   B.1.617.1  Furthermore, among these lineages, WHO identif...\n",
       "86   PMC8754802     B.1.525  Furthermore, among these lineages, WHO identif...\n",
       "87   PMC8754802   B.1.617.2  Furthermore, among these lineages, WHO identif...\n",
       "88   PMC8754802     B.1.526  Furthermore, among these lineages, WHO identif...\n",
       "89   PMC8754802        C.37  Furthermore, among these lineages, WHO identif...\n",
       "90   PMC8754802     B.1.1.7  Furthermore, among these lineages, WHO identif...\n",
       "91   PMC8754802         P.2  Furthermore, among these lineages, WHO identif...\n",
       "92   PMC8754802   B.1.214.2  For the rest of lineages, we identified indel ...\n",
       "93   PMC8754802       A.2.5  For the rest of lineages, we identified indel ...\n",
       "94   PMC8754802      B.1.36  For the rest of lineages, we identified indel ...\n",
       "95   PMC8707919   B.1.617.2  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "96   PMC8707919   B.1.617.1  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "97   PMC8707919     B.1.617  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "98   PMC8707919         P.1  In comparison, similar modest reductions were ...\n",
       "99   PMC8707919     B.1.526  In comparison, similar modest reductions were ...\n",
       "100  PMC8707919        C.37  In comparison, similar modest reductions were ...\n",
       "101  PMC8707919         R.1  In comparison, similar modest reductions were ...\n",
       "102  PMC8707919     B.1.351  In comparison, similar modest reductions were ...\n",
       "103  PMC8707919         B.1  One of the earliest variants that is highly in...\n",
       "104  PMC8707919     B.1.429  Subsequent genomic surveillance has led to the...\n",
       "105  PMC8707919     B.1.427  The B.1.427/B.1.429 and B.1.617 lineage varian...\n",
       "106  PMC8707919         P.3  The E484K substitution found in B.1.351, B.1.5...\n",
       "107  PMC8707919     B.1.1.7  These findings extend a prior study reporting ...\n",
       "108  PMC8707919       B.1.2  Furthermore, continued surveillance is necessa...\n",
       "109  PMC8707919       C.1.2  Furthermore, continued surveillance is necessa...\n",
       "110  PMC8707919     B.1.525  Furthermore, continued surveillance is necessa...\n",
       "111  PMC8707919   B.1.617.3  Sequences corresponding to the S1/S2 region of...\n",
       "112   PPR428671         P.1  Genomic surveillance of SC2 variants has been ...\n",
       "113   PPR428671       A.2.5  Here, we describe the introduction of SC2 line...\n",
       "114   PPR428671       A.2.4  In January 2021, BPHL detected the first cases...\n",
       "115   PPR428671         B.1  The A.2.5 lineage in Florida became of particu...\n",
       "116   PPR428671     B.1.427  A.2.5 also contains spike L452R, a CDC-defined...\n",
       "117   PPR428671   B.1.617.2  Allele L452R has independently emerged in seve...\n",
       "118   PPR428671     B.1.526  Allele L452R has independently emerged in seve...\n",
       "119   PPR428671   B.1.617.1  Allele L452R has independently emerged in seve...\n",
       "120   PPR428671        C.36  Allele L452R has independently emerged in seve...\n",
       "121   PPR428671     B.1.1.7  Additional mutations causing increased transmi...\n",
       "122  PMC8635972     B.1.1.7  The alpha/B.1.1.7 SARS-CoV-2 lineage emerged i...\n",
       "123  PMC8635972         P.1  The incidence domination was likely due to a f...\n",
       "124  PMC8635972     B.1.351  The incidence domination was likely due to a f...\n",
       "125  PMC8635972     B.1.621  Some of these residue substitutions, including...\n",
       "126  PMC8635972         P.3  Some of these residue substitutions, including...\n",
       "127  PMC8635972   B.1.617.2  All the above VOC/VOI have been for the most p...\n",
       "128  PMC8635972         B.1  In order to examine whether the increased affi...\n",
       "129  PMC8646660     B.1.1.7  FD20 engages the RBD at an epitope distal to t...\n",
       "130  PMC8646660   B.1.617.2  FD20 engages the RBD at an epitope distal to t...\n",
       "131  PMC8646660         P.1  FD20 engages the RBD at an epitope distal to t...\n",
       "132  PMC8646660     B.1.351  FD20 engages the RBD at an epitope distal to t...\n",
       "133  PMC8646660   B.1.1.298  Y453F increases the ACE2 binding affinity, and...\n",
       "134  PMC8646660     B.1.177  A222V and S477N are described in 20A.EU1 (B.1....\n",
       "135  PMC8646660     B.1.160  A222V and S477N are described in 20A.EU1 (B.1....\n",
       "136  PMC8646660     B.1.141  N439K is in B.1.141, B.1.258, and in mink stra...\n",
       "137  PMC8646660     B.1.258  N439K is in B.1.141, B.1.258, and in mink stra...\n",
       "138  PMC8602852     B.1.621  Four SARS-CoV-2 variants of concern including ...\n",
       "139  PMC8602852   B.1.617.2  Four SARS-CoV-2 variants of concern including ...\n",
       "140  PMC8602852         P.1  Four SARS-CoV-2 variants of concern including ...\n",
       "141  PMC8602852        C.37  Four SARS-CoV-2 variants of concern including ...\n",
       "142  PMC8602852     B.1.1.7  Four SARS-CoV-2 variants of concern including ...\n",
       "143  PMC8602852     B.1.351  Four SARS-CoV-2 variants of concern including ...\n",
       "144  PMC8602852     B.1.619  Currently designated VUMs include B.1.427/B.1....\n",
       "145  PMC8602852   B.1.214.2  Currently designated VUMs include B.1.427/B.1....\n",
       "146  PMC8602852     B.1.427  Currently designated VUMs include B.1.427/B.1....\n",
       "147  PMC8602852     B.1.526  Currently designated VUMs include B.1.427/B.1....\n",
       "148  PMC8602852   B.1.466.2  Currently designated VUMs include B.1.427/B.1....\n",
       "149  PMC8602852     B.1.620  Currently designated VUMs include B.1.427/B.1....\n",
       "150  PMC8602852   B.1.1.523  Currently designated VUMs include B.1.427/B.1....\n",
       "151  PMC8602852   B.1.1.318  Currently designated VUMs include B.1.427/B.1....\n",
       "152  PMC8602852   B.1.1.519  Currently designated VUMs include B.1.427/B.1....\n",
       "153  PMC8602852   B.1.617.1  Currently designated VUMs include B.1.427/B.1....\n",
       "154  PMC8602852         R.1  Currently designated VUMs include B.1.427/B.1....\n",
       "155  PMC8602852       C.1.2  Currently designated VUMs include B.1.427/B.1....\n",
       "156  PMC8602852     B.1.525  Currently designated VUMs include B.1.427/B.1....\n",
       "157  PMC8602852      C.36.3  Currently designated VUMs include B.1.427/B.1....\n",
       "158  PMC8602852         P.3  Some former VOIs but no longer designated as V...\n",
       "159  PMC8602852         P.2  Some former VOIs but no longer designated as V...\n",
       "160  PMC8602852    B.1.1.28  P.3 as well as P.1 belong to B.1.1.28 lineage ...\n",
       "161  PMC8602852     B.1.617  At the same time, B.1.525 was resistant to neu...\n",
       "162  PMC8602852   B.1.617.3  With the spread of the local epidemic in India...\n",
       "163  PMC8598544   B.1.617.2  The Delta (B.1.617.2) variant was first identi...\n",
       "164  PMC8598544        C.37  The Lambda variant belongs to the C.37 lineage...\n",
       "165  PMC8598544     B.1.621  On August 30, 2021, the WHO announced the Mu v...\n",
       "166  PMC8535385   B.1.617.2  The new variants are accumulating mutations in...\n",
       "167  PMC8535385         P.1  The new variants are accumulating mutations in...\n",
       "168  PMC8535385     B.1.1.7  The new variants are accumulating mutations in...\n",
       "169  PMC8535385     B.1.351  The new variants are accumulating mutations in...\n",
       "170  PMC8501352     B.1.117  At the time of the first clinical trials the “...\n",
       "171  PMC8525575     B.1.427  The massive emergence of L452R variants was fi...\n",
       "172  PMC8525575     B.1.232  By PCR amplification and Sanger sequencing of ...\n",
       "173  PMC8525575    B.1.1.28  Because of the important functional and antige...\n",
       "174  PMC8525575     B.1.1.7  Because of the important functional and antige...\n",
       "175  PMC8525575     B.1.351  Because of the important functional and antige...\n",
       "176  PMC8525575        A.27  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "177  PMC8525575   B.1.617.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "178  PMC8525575   B.1.1.263  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "179  PMC8525575      B.1.36  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "180  PMC8525575        A.21  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "181  PMC8525575     B.1.466  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "182  PMC8525575   B.1.1.459  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "183  PMC8525575   B.1.526.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "184  PMC8525575     A.2.5.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "185  PMC8525575     A.2.5.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "186  PMC8525575   B.1.617.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "187  PMC8525575     B.1.429  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "188  PMC8525575  B.1.177.83  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "189  PMC8525575      C.36.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "190  PMC8525575   B.1.617.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "191  PMC8525575         P.4  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "192  PMC8525575        C.37  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "193  PMC8525575        C.16  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "194  PMC8525575      C.36.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "195   PPR380399         P.1  In late 2020, three SARS-CoV-2 variants of con...\n",
       "196   PPR380399     B.1.1.7  In late 2020, three SARS-CoV-2 variants of con...\n",
       "197   PPR380399     B.1.351  In late 2020, three SARS-CoV-2 variants of con...\n",
       "198   PPR380399     B.1.427  In early 2021, several new VOC appeared includ...\n",
       "199   PPR380399     B.1.526  In early 2021, several new VOC appeared includ...\n",
       "200   PPR380399     B.1.617  In early 2021, several new VOC appeared includ...\n",
       "201   PPR380399         D.2  The D.2 variant in Australia was defined by an...\n",
       "202   PPR380399     B.1.160  In contrast, B.1.160, which carries S S477N so...\n",
       "203   PPR380399     B.1.429  The substitution D260Y, present in B.1.427 and...\n",
       "204   PPR380399     B.1.177  Several lineages show introgression from other...\n",
       "205   PPR380399         B.1  c. Population-level analysis of new mutations ...\n",
       "206  PMC7705438         V.2  Variants with minor allele frequency less than...\n",
       "207  PMC7705438        C.14  We identified one previously reported (c.2729 ...\n",
       "208  PMC7705438         V.1  All variants were reclassified according to AC...\n",
       "209  PMC7705438        C.38  Results: We found 8 mutations in this patient ..."
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "from        object\n",
       "to          object\n",
       "evidence    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "StringArray requires a sequence of strings or pandas.NA",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-154-cd258e7f3942>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msub\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'doi'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'url'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'pubmedId'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'journal'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'year'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'title'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'abstract'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'string'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtypes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   5691\u001b[0m             results = [\n\u001b[1;32m   5692\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5693\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5694\u001b[0m             ]\n\u001b[1;32m   5695\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   5691\u001b[0m             results = [\n\u001b[1;32m   5692\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5693\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5694\u001b[0m             ]\n\u001b[1;32m   5695\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   5696\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5697\u001b[0m             \u001b[0;31m# else, only a single dtype is given\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5698\u001b[0;31m             \u001b[0mnew_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5699\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5700\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    580\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    581\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"raise\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 582\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"astype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    583\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    584\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, f, filter, **kwargs)\u001b[0m\n\u001b[1;32m    440\u001b[0m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 442\u001b[0;31m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    443\u001b[0m             \u001b[0mresult_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_extend_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mapplied\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult_blocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    444\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    623\u001b[0m             \u001b[0mvals1d\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 625\u001b[0;31m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mastype_nansafe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvals1d\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    626\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mValueError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m                 \u001b[0;31m# e.g. astype_nansafe can fail on object-dtype of strings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/cast.py\u001b[0m in \u001b[0;36mastype_nansafe\u001b[0;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[1;32m    819\u001b[0m     \u001b[0;31m# dispatch on extension dtype if needed\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    820\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mis_extension_array_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 821\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconstruct_array_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_from_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    822\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    823\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/arrays/string_.py\u001b[0m in \u001b[0;36m_from_sequence\u001b[0;34m(cls, scalars, dtype, copy)\u001b[0m\n\u001b[1;32m    195\u001b[0m             \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mna_values\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mStringDtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mna_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    196\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 197\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    198\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    199\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/arrays/string_.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, values, copy)\u001b[0m\n\u001b[1;32m    164\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mStringDtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mskip_validation\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.7/site-packages/pandas/core/arrays/string_.py\u001b[0m in \u001b[0;36m_validate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    169\u001b[0m         \u001b[0;34m\"\"\"Validate that we only store NA or strings.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ndarray\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_string_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ndarray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskipna\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 171\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"StringArray requires a sequence of strings or pandas.NA\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    172\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ndarray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"object\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    173\u001b[0m             raise ValueError(\n",
      "\u001b[0;31mValueError\u001b[0m: StringArray requires a sequence of strings or pandas.NA"
     ]
    }
   ],
   "source": [
    "sub[['id', 'name', 'doi', 'url', 'pubmedId', 'pmcId', 'journal', 'year','title', 'abstract']].astype('string').dtypes"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "a = a[['ID','lineage','string']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Since the original strain of severe acute resp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>Q.3</td>\n",
       "      <td>Among the 732 clinical samples that were teste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PMC9119066</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>By WGS, these samples were variant of interest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>This strain clade 201/501Y.V1, Pango lineage B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>26\\n There are three mutations of significant ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>P.1</td>\n",
       "      <td>56\\n Safety and immunogenicity study of a SARS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Gamma variant is particularly found from the o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>These resembling variants, B.1.427 and B.1.429...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>These resembling variants, B.1.427 and B.1.429...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>B.1.617.2 (Delta variant) is a variant of line...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>83\\n The B.1.617.1 strain is 6.8‐fold more res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>92\\n Researchers discovered that this new C.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PMC9088647</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>K417N, N440K, G446S, S477N, T478K, E484A, Q493...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The three largest cities in Colombia: Bogotá, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.625</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.348</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>C.37.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>C.37</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.1.487</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1.621.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>P.1.10</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PMC9008484</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The figure shows the number of cases for SARS-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>P.1</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>We also look at the effectiveness of approved ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>The D614G substitution is found in most of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The D614G substitution is found in most of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>In addition, some of the notable substitutions...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>In addition, some of the notable substitutions...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>Fitness-enhancing mutations have been reported...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>However, in-vitro neutralization assays show t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PMC8902153</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>According to the CDC, US, the P.1 variant seem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Since the original strain of SARS-CoV-2 virus ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>P.1</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>*), Beta (B.1.351), Gamma (P.1 and sublineages...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>*) and Omicron (B.1.1.529 and BA.*).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>Q.3</td>\n",
       "      <td>Among the 732 clinical samples that were teste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>By WGS, these samples were variant of interest...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>PPR452207</td>\n",
       "      <td>B.1.1.519</td>\n",
       "      <td>*), and Omicron (B.1.1.519 and BA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>P.1</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>C.37</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>As of December 2021, the World Health Organiza...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>PMC8942852</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>On 24 November 2021, the B.1.1.529 lineage, a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>P.1</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>A total of 411 samples belonging to the Alpha ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The second largest clade identified by Nextcla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.369</td>\n",
       "      <td>The second largest clade identified by Nextcla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>Pangolin classified most of these as the US st...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>A clear trend can be observed with the average...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.324</td>\n",
       "      <td>While this mutation was observed in the B.1.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>PMC8769358</td>\n",
       "      <td>B.1.1.317</td>\n",
       "      <td>The A845S spike mutation was observed at the s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>P.1</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>As a result of viral evolution, the initial SA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.1.529</td>\n",
       "      <td>The most recent VOC, first identified in South...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.529</td>\n",
       "      <td>Their neutralization spectrum is different and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>PPR442789</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>Residues mutated in B.1.1.7, B.1.617 and B.1.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Although neutralization of one of these antibo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Although neutralization of one of these antibo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Finally, we investigate the effects of B.1.1.7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>PMC8690919</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Similar histopathological findings of lung and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>C.37</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Furthermore, among these lineages, WHO identif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.214.2</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PMC8754802</td>\n",
       "      <td>B.1.36</td>\n",
       "      <td>For the rest of lineages, we identified indel ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>The SARS-CoV-2 B.1.617 lineage variants, Kappa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>P.1</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>C.37</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>R.1</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>In comparison, similar modest reductions were ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1</td>\n",
       "      <td>One of the earliest variants that is highly in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>Subsequent genomic surveillance has led to the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>The B.1.427/B.1.429 and B.1.617 lineage varian...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>P.3</td>\n",
       "      <td>The E484K substitution found in B.1.351, B.1.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>These findings extend a prior study reporting ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Furthermore, continued surveillance is necessa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PMC8707919</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>Sequences corresponding to the S1/S2 region of...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Genomic surveillance of SC2 variants has been ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>A.2.5</td>\n",
       "      <td>Here, we describe the introduction of SC2 line...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>A.2.4</td>\n",
       "      <td>In January 2021, BPHL detected the first cases...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1</td>\n",
       "      <td>The A.2.5 lineage in Florida became of particu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>A.2.5 also contains spike L452R, a CDC-defined...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>C.36</td>\n",
       "      <td>Allele L452R has independently emerged in seve...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PPR428671</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Additional mutations causing increased transmi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The alpha/B.1.1.7 SARS-CoV-2 lineage emerged i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The incidence domination was likely due to a f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>The incidence domination was likely due to a f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Some of these residue substitutions, including...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Some of these residue substitutions, including...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>All the above VOC/VOI have been for the most p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PMC8635972</td>\n",
       "      <td>B.1</td>\n",
       "      <td>In order to examine whether the increased affi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>P.1</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>FD20 engages the RBD at an epitope distal to t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.1.298</td>\n",
       "      <td>Y453F increases the ACE2 binding affinity, and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.177</td>\n",
       "      <td>A222V and S477N are described in 20A.EU1 (B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.160</td>\n",
       "      <td>A222V and S477N are described in 20A.EU1 (B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.141</td>\n",
       "      <td>N439K is in B.1.141, B.1.258, and in mink stra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PMC8646660</td>\n",
       "      <td>B.1.258</td>\n",
       "      <td>N439K is in B.1.141, B.1.258, and in mink stra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.1</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.37</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Four SARS-CoV-2 variants of concern including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.619</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.214.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.466.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.620</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.523</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.519</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>R.1</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.1.2</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>C.36.3</td>\n",
       "      <td>Currently designated VUMs include B.1.427/B.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.3</td>\n",
       "      <td>Some former VOIs but no longer designated as V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>P.2</td>\n",
       "      <td>Some former VOIs but no longer designated as V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>P.3 as well as P.1 belong to B.1.1.28 lineage ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>At the same time, B.1.525 was resistant to neu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>PMC8602852</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>With the spread of the local epidemic in India...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The Delta (B.1.617.2) variant was first identi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>C.37</td>\n",
       "      <td>The Lambda variant belongs to the C.37 lineage...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>PMC8598544</td>\n",
       "      <td>B.1.621</td>\n",
       "      <td>On August 30, 2021, the WHO announced the Mu v...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>P.1</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>PMC8535385</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>The new variants are accumulating mutations in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>PMC8501352</td>\n",
       "      <td>B.1.117</td>\n",
       "      <td>At the time of the first clinical trials the “...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>The massive emergence of L452R variants was fi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.232</td>\n",
       "      <td>By PCR amplification and Sanger sequencing of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Because of the important functional and antige...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.27</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.3</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.263</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.36</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.21</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.466</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.1.459</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.526.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.2.5.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>A.2.5.2</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.177.83</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.36.1</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>P.4</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.37</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.16</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PMC8525575</td>\n",
       "      <td>C.36.3</td>\n",
       "      <td>A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>P.1</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>In late 2020, three SARS-CoV-2 variants of con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.427</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.526</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.617</td>\n",
       "      <td>In early 2021, several new VOC appeared includ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>D.2</td>\n",
       "      <td>The D.2 variant in Australia was defined by an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.160</td>\n",
       "      <td>In contrast, B.1.160, which carries S S477N so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.429</td>\n",
       "      <td>The substitution D260Y, present in B.1.427 and...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1.177</td>\n",
       "      <td>Several lineages show introgression from other...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>PPR380399</td>\n",
       "      <td>B.1</td>\n",
       "      <td>c. Population-level analysis of new mutations ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>V.2</td>\n",
       "      <td>Variants with minor allele frequency less than...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>C.14</td>\n",
       "      <td>We identified one previously reported (c.2729 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>V.1</td>\n",
       "      <td>All variants were reclassified according to AC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>PMC7705438</td>\n",
       "      <td>C.38</td>\n",
       "      <td>Results: We found 8 mutations in this patient ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           from          to                                           evidence\n",
       "0    PMC9119066   B.1.1.529  Since the original strain of severe acute resp...\n",
       "1    PMC9119066   B.1.617.2  Since the original strain of severe acute resp...\n",
       "2    PMC9119066         P.1  Since the original strain of severe acute resp...\n",
       "3    PMC9119066     B.1.1.7  Since the original strain of severe acute resp...\n",
       "4    PMC9119066     B.1.351  Since the original strain of severe acute resp...\n",
       "5    PMC9119066         Q.3  Among the 732 clinical samples that were teste...\n",
       "6    PMC9119066     B.1.621  By WGS, these samples were variant of interest...\n",
       "7    PMC9088647     B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...\n",
       "8    PMC9088647     B.1.351  26\\n There are three mutations of significant ...\n",
       "9    PMC9088647         P.1  56\\n Safety and immunogenicity study of a SARS...\n",
       "10   PMC9088647         P.2  Gamma variant is particularly found from the o...\n",
       "11   PMC9088647     B.1.427  These resembling variants, B.1.427 and B.1.429...\n",
       "12   PMC9088647     B.1.429  These resembling variants, B.1.427 and B.1.429...\n",
       "13   PMC9088647     B.1.617  B.1.617.2 (Delta variant) is a variant of line...\n",
       "14   PMC9088647   B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...\n",
       "15   PMC9088647       C.1.2  92\\n Researchers discovered that this new C.1....\n",
       "16   PMC9088647   B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...\n",
       "17   PMC9008484     B.1.621  The three largest cities in Colombia: Bogotá, ...\n",
       "18   PMC9008484   B.1.617.2  The three largest cities in Colombia: Bogotá, ...\n",
       "19   PMC9008484         P.1  The three largest cities in Colombia: Bogotá, ...\n",
       "20   PMC9008484     B.1.1.7  The three largest cities in Colombia: Bogotá, ...\n",
       "21   PMC9008484       B.1.1  The figure shows the number of cases for SARS-...\n",
       "22   PMC9008484       A.2.5  The figure shows the number of cases for SARS-...\n",
       "23   PMC9008484     B.1.625  The figure shows the number of cases for SARS-...\n",
       "24   PMC9008484   B.1.1.348  The figure shows the number of cases for SARS-...\n",
       "25   PMC9008484      C.37.1  The figure shows the number of cases for SARS-...\n",
       "26   PMC9008484        C.37  The figure shows the number of cases for SARS-...\n",
       "27   PMC9008484   B.1.1.487  The figure shows the number of cases for SARS-...\n",
       "28   PMC9008484   B.1.621.1  The figure shows the number of cases for SARS-...\n",
       "29   PMC9008484      P.1.10  The figure shows the number of cases for SARS-...\n",
       "30   PMC9008484         B.1  The figure shows the number of cases for SARS-...\n",
       "31   PMC8902153   B.1.617.2  We also look at the effectiveness of approved ...\n",
       "32   PMC8902153         P.1  We also look at the effectiveness of approved ...\n",
       "33   PMC8902153     B.1.351  We also look at the effectiveness of approved ...\n",
       "34   PMC8902153   B.1.1.529  The D614G substitution is found in most of the...\n",
       "35   PMC8902153     B.1.1.7  The D614G substitution is found in most of the...\n",
       "36   PMC8902153     B.1.617  In addition, some of the notable substitutions...\n",
       "37   PMC8902153     B.1.429  In addition, some of the notable substitutions...\n",
       "38   PMC8902153       B.1.1  Fitness-enhancing mutations have been reported...\n",
       "39   PMC8902153   B.1.617.1  However, in-vitro neutralization assays show t...\n",
       "40   PMC8902153     B.1.427  According to the CDC, US, the P.1 variant seem...\n",
       "41    PPR452207     B.1.1.7  Since the original strain of SARS-CoV-2 virus ...\n",
       "42    PPR452207   B.1.617.2  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "43    PPR452207         P.1  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "44    PPR452207     B.1.351  *), Beta (B.1.351), Gamma (P.1 and sublineages...\n",
       "45    PPR452207   B.1.1.529               *) and Omicron (B.1.1.529 and BA.*).\n",
       "46    PPR452207         Q.3  Among the 732 clinical samples that were teste...\n",
       "47    PPR452207     B.1.621  By WGS, these samples were variant of interest...\n",
       "48    PPR452207   B.1.1.519                 *), and Omicron (B.1.1.519 and BA.\n",
       "49   PMC8942852     B.1.621  As of December 2021, the World Health Organiza...\n",
       "50   PMC8942852   B.1.1.529  As of December 2021, the World Health Organiza...\n",
       "51   PMC8942852   B.1.617.2  As of December 2021, the World Health Organiza...\n",
       "52   PMC8942852         P.1  As of December 2021, the World Health Organiza...\n",
       "53   PMC8942852        C.37  As of December 2021, the World Health Organiza...\n",
       "54   PMC8942852     B.1.1.7  As of December 2021, the World Health Organiza...\n",
       "55   PMC8942852     B.1.351  As of December 2021, the World Health Organiza...\n",
       "56   PMC8942852       B.1.1  On 24 November 2021, the B.1.1.529 lineage, a ...\n",
       "57   PMC8769358     B.1.427  A total of 411 samples belonging to the Alpha ...\n",
       "58   PMC8769358   B.1.617.2  A total of 411 samples belonging to the Alpha ...\n",
       "59   PMC8769358         P.1  A total of 411 samples belonging to the Alpha ...\n",
       "60   PMC8769358     B.1.1.7  A total of 411 samples belonging to the Alpha ...\n",
       "61   PMC8769358     B.1.351  A total of 411 samples belonging to the Alpha ...\n",
       "62   PMC8769358         B.1  The second largest clade identified by Nextcla...\n",
       "63   PMC8769358     B.1.369  The second largest clade identified by Nextcla...\n",
       "64   PMC8769358       B.1.2  Pangolin classified most of these as the US st...\n",
       "65   PMC8769358     B.1.526  A clear trend can be observed with the average...\n",
       "66   PMC8769358       B.1.1  While this mutation was observed in the B.1.1....\n",
       "67   PMC8769358     B.1.621  While this mutation was observed in the B.1.1....\n",
       "68   PMC8769358     B.1.324  While this mutation was observed in the B.1.1....\n",
       "69   PMC8769358   B.1.1.317  The A845S spike mutation was observed at the s...\n",
       "70    PPR442789   B.1.617.2  As a result of viral evolution, the initial SA...\n",
       "71    PPR442789         P.1  As a result of viral evolution, the initial SA...\n",
       "72    PPR442789     B.1.1.7  As a result of viral evolution, the initial SA...\n",
       "73    PPR442789     B.1.351  As a result of viral evolution, the initial SA...\n",
       "74    PPR442789   B.1.1.529  The most recent VOC, first identified in South...\n",
       "75    PPR442789     B.1.529  Their neutralization spectrum is different and...\n",
       "76    PPR442789     B.1.617  Residues mutated in B.1.1.7, B.1.617 and B.1.3...\n",
       "77   PMC8690919     B.1.1.7  Although neutralization of one of these antibo...\n",
       "78   PMC8690919     B.1.351  Although neutralization of one of these antibo...\n",
       "79   PMC8690919   B.1.617.2  Finally, we investigate the effects of B.1.1.7...\n",
       "80   PMC8690919         P.1  Similar histopathological findings of lung and...\n",
       "81   PMC8754802         P.1  Furthermore, among these lineages, WHO identif...\n",
       "82   PMC8754802         P.3  Furthermore, among these lineages, WHO identif...\n",
       "83   PMC8754802     B.1.351  Furthermore, among these lineages, WHO identif...\n",
       "84   PMC8754802     B.1.621  Furthermore, among these lineages, WHO identif...\n",
       "85   PMC8754802   B.1.617.1  Furthermore, among these lineages, WHO identif...\n",
       "86   PMC8754802     B.1.525  Furthermore, among these lineages, WHO identif...\n",
       "87   PMC8754802   B.1.617.2  Furthermore, among these lineages, WHO identif...\n",
       "88   PMC8754802     B.1.526  Furthermore, among these lineages, WHO identif...\n",
       "89   PMC8754802        C.37  Furthermore, among these lineages, WHO identif...\n",
       "90   PMC8754802     B.1.1.7  Furthermore, among these lineages, WHO identif...\n",
       "91   PMC8754802         P.2  Furthermore, among these lineages, WHO identif...\n",
       "92   PMC8754802   B.1.214.2  For the rest of lineages, we identified indel ...\n",
       "93   PMC8754802       A.2.5  For the rest of lineages, we identified indel ...\n",
       "94   PMC8754802      B.1.36  For the rest of lineages, we identified indel ...\n",
       "95   PMC8707919   B.1.617.2  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "96   PMC8707919   B.1.617.1  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "97   PMC8707919     B.1.617  The SARS-CoV-2 B.1.617 lineage variants, Kappa...\n",
       "98   PMC8707919         P.1  In comparison, similar modest reductions were ...\n",
       "99   PMC8707919     B.1.526  In comparison, similar modest reductions were ...\n",
       "100  PMC8707919        C.37  In comparison, similar modest reductions were ...\n",
       "101  PMC8707919         R.1  In comparison, similar modest reductions were ...\n",
       "102  PMC8707919     B.1.351  In comparison, similar modest reductions were ...\n",
       "103  PMC8707919         B.1  One of the earliest variants that is highly in...\n",
       "104  PMC8707919     B.1.429  Subsequent genomic surveillance has led to the...\n",
       "105  PMC8707919     B.1.427  The B.1.427/B.1.429 and B.1.617 lineage varian...\n",
       "106  PMC8707919         P.3  The E484K substitution found in B.1.351, B.1.5...\n",
       "107  PMC8707919     B.1.1.7  These findings extend a prior study reporting ...\n",
       "108  PMC8707919       B.1.2  Furthermore, continued surveillance is necessa...\n",
       "109  PMC8707919       C.1.2  Furthermore, continued surveillance is necessa...\n",
       "110  PMC8707919     B.1.525  Furthermore, continued surveillance is necessa...\n",
       "111  PMC8707919   B.1.617.3  Sequences corresponding to the S1/S2 region of...\n",
       "112   PPR428671         P.1  Genomic surveillance of SC2 variants has been ...\n",
       "113   PPR428671       A.2.5  Here, we describe the introduction of SC2 line...\n",
       "114   PPR428671       A.2.4  In January 2021, BPHL detected the first cases...\n",
       "115   PPR428671         B.1  The A.2.5 lineage in Florida became of particu...\n",
       "116   PPR428671     B.1.427  A.2.5 also contains spike L452R, a CDC-defined...\n",
       "117   PPR428671   B.1.617.2  Allele L452R has independently emerged in seve...\n",
       "118   PPR428671     B.1.526  Allele L452R has independently emerged in seve...\n",
       "119   PPR428671   B.1.617.1  Allele L452R has independently emerged in seve...\n",
       "120   PPR428671        C.36  Allele L452R has independently emerged in seve...\n",
       "121   PPR428671     B.1.1.7  Additional mutations causing increased transmi...\n",
       "122  PMC8635972     B.1.1.7  The alpha/B.1.1.7 SARS-CoV-2 lineage emerged i...\n",
       "123  PMC8635972         P.1  The incidence domination was likely due to a f...\n",
       "124  PMC8635972     B.1.351  The incidence domination was likely due to a f...\n",
       "125  PMC8635972     B.1.621  Some of these residue substitutions, including...\n",
       "126  PMC8635972         P.3  Some of these residue substitutions, including...\n",
       "127  PMC8635972   B.1.617.2  All the above VOC/VOI have been for the most p...\n",
       "128  PMC8635972         B.1  In order to examine whether the increased affi...\n",
       "129  PMC8646660     B.1.1.7  FD20 engages the RBD at an epitope distal to t...\n",
       "130  PMC8646660   B.1.617.2  FD20 engages the RBD at an epitope distal to t...\n",
       "131  PMC8646660         P.1  FD20 engages the RBD at an epitope distal to t...\n",
       "132  PMC8646660     B.1.351  FD20 engages the RBD at an epitope distal to t...\n",
       "133  PMC8646660   B.1.1.298  Y453F increases the ACE2 binding affinity, and...\n",
       "134  PMC8646660     B.1.177  A222V and S477N are described in 20A.EU1 (B.1....\n",
       "135  PMC8646660     B.1.160  A222V and S477N are described in 20A.EU1 (B.1....\n",
       "136  PMC8646660     B.1.141  N439K is in B.1.141, B.1.258, and in mink stra...\n",
       "137  PMC8646660     B.1.258  N439K is in B.1.141, B.1.258, and in mink stra...\n",
       "138  PMC8602852     B.1.621  Four SARS-CoV-2 variants of concern including ...\n",
       "139  PMC8602852   B.1.617.2  Four SARS-CoV-2 variants of concern including ...\n",
       "140  PMC8602852         P.1  Four SARS-CoV-2 variants of concern including ...\n",
       "141  PMC8602852        C.37  Four SARS-CoV-2 variants of concern including ...\n",
       "142  PMC8602852     B.1.1.7  Four SARS-CoV-2 variants of concern including ...\n",
       "143  PMC8602852     B.1.351  Four SARS-CoV-2 variants of concern including ...\n",
       "144  PMC8602852     B.1.619  Currently designated VUMs include B.1.427/B.1....\n",
       "145  PMC8602852   B.1.214.2  Currently designated VUMs include B.1.427/B.1....\n",
       "146  PMC8602852     B.1.427  Currently designated VUMs include B.1.427/B.1....\n",
       "147  PMC8602852     B.1.526  Currently designated VUMs include B.1.427/B.1....\n",
       "148  PMC8602852   B.1.466.2  Currently designated VUMs include B.1.427/B.1....\n",
       "149  PMC8602852     B.1.620  Currently designated VUMs include B.1.427/B.1....\n",
       "150  PMC8602852   B.1.1.523  Currently designated VUMs include B.1.427/B.1....\n",
       "151  PMC8602852   B.1.1.318  Currently designated VUMs include B.1.427/B.1....\n",
       "152  PMC8602852   B.1.1.519  Currently designated VUMs include B.1.427/B.1....\n",
       "153  PMC8602852   B.1.617.1  Currently designated VUMs include B.1.427/B.1....\n",
       "154  PMC8602852         R.1  Currently designated VUMs include B.1.427/B.1....\n",
       "155  PMC8602852       C.1.2  Currently designated VUMs include B.1.427/B.1....\n",
       "156  PMC8602852     B.1.525  Currently designated VUMs include B.1.427/B.1....\n",
       "157  PMC8602852      C.36.3  Currently designated VUMs include B.1.427/B.1....\n",
       "158  PMC8602852         P.3  Some former VOIs but no longer designated as V...\n",
       "159  PMC8602852         P.2  Some former VOIs but no longer designated as V...\n",
       "160  PMC8602852    B.1.1.28  P.3 as well as P.1 belong to B.1.1.28 lineage ...\n",
       "161  PMC8602852     B.1.617  At the same time, B.1.525 was resistant to neu...\n",
       "162  PMC8602852   B.1.617.3  With the spread of the local epidemic in India...\n",
       "163  PMC8598544   B.1.617.2  The Delta (B.1.617.2) variant was first identi...\n",
       "164  PMC8598544        C.37  The Lambda variant belongs to the C.37 lineage...\n",
       "165  PMC8598544     B.1.621  On August 30, 2021, the WHO announced the Mu v...\n",
       "166  PMC8535385   B.1.617.2  The new variants are accumulating mutations in...\n",
       "167  PMC8535385         P.1  The new variants are accumulating mutations in...\n",
       "168  PMC8535385     B.1.1.7  The new variants are accumulating mutations in...\n",
       "169  PMC8535385     B.1.351  The new variants are accumulating mutations in...\n",
       "170  PMC8501352     B.1.117  At the time of the first clinical trials the “...\n",
       "171  PMC8525575     B.1.427  The massive emergence of L452R variants was fi...\n",
       "172  PMC8525575     B.1.232  By PCR amplification and Sanger sequencing of ...\n",
       "173  PMC8525575    B.1.1.28  Because of the important functional and antige...\n",
       "174  PMC8525575     B.1.1.7  Because of the important functional and antige...\n",
       "175  PMC8525575     B.1.351  Because of the important functional and antige...\n",
       "176  PMC8525575        A.27  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "177  PMC8525575   B.1.617.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "178  PMC8525575   B.1.1.263  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "179  PMC8525575      B.1.36  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "180  PMC8525575        A.21  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "181  PMC8525575     B.1.466  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "182  PMC8525575   B.1.1.459  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "183  PMC8525575   B.1.526.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "184  PMC8525575     A.2.5.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "185  PMC8525575     A.2.5.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "186  PMC8525575   B.1.617.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "187  PMC8525575     B.1.429  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "188  PMC8525575  B.1.177.83  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "189  PMC8525575      C.36.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "190  PMC8525575   B.1.617.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "191  PMC8525575         P.4  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "192  PMC8525575        C.37  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "193  PMC8525575        C.16  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "194  PMC8525575      C.36.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...\n",
       "195   PPR380399         P.1  In late 2020, three SARS-CoV-2 variants of con...\n",
       "196   PPR380399     B.1.1.7  In late 2020, three SARS-CoV-2 variants of con...\n",
       "197   PPR380399     B.1.351  In late 2020, three SARS-CoV-2 variants of con...\n",
       "198   PPR380399     B.1.427  In early 2021, several new VOC appeared includ...\n",
       "199   PPR380399     B.1.526  In early 2021, several new VOC appeared includ...\n",
       "200   PPR380399     B.1.617  In early 2021, several new VOC appeared includ...\n",
       "201   PPR380399         D.2  The D.2 variant in Australia was defined by an...\n",
       "202   PPR380399     B.1.160  In contrast, B.1.160, which carries S S477N so...\n",
       "203   PPR380399     B.1.429  The substitution D260Y, present in B.1.427 and...\n",
       "204   PPR380399     B.1.177  Several lineages show introgression from other...\n",
       "205   PPR380399         B.1  c. Population-level analysis of new mutations ...\n",
       "206  PMC7705438         V.2  Variants with minor allele frequency less than...\n",
       "207  PMC7705438        C.14  We identified one previously reported (c.2729 ...\n",
       "208  PMC7705438         V.1  All variants were reclassified according to AC...\n",
       "209  PMC7705438        C.38  Results: We found 8 mutations in this patient ..."
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.columns = ['from','to','evidence']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "a.to_csv('Publication-MENTIONS-Lineage.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get PANGO lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pango = pd.read_csv(NEO4J_IMPORT + \"/00b-PANGOLineage.csv\", dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>alias</th>\n",
       "      <th>predecessor</th>\n",
       "      <th>l0</th>\n",
       "      <th>l1</th>\n",
       "      <th>l2</th>\n",
       "      <th>l3</th>\n",
       "      <th>levels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>B.1.418</td>\n",
       "      <td>Scottish lineage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B.1.418</td>\n",
       "      <td>B.1</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1327</th>\n",
       "      <td>B.1.571</td>\n",
       "      <td>USA lineage, split out from B.1.369</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B.1.571</td>\n",
       "      <td>B.1</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>B.1.286</td>\n",
       "      <td>Withdrawn: Reassigned in the current tree. USA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B.1.286</td>\n",
       "      <td>B.1</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1622</th>\n",
       "      <td>AY.131</td>\n",
       "      <td>Alias of B.1.617.2.131, mainly reported from B...</td>\n",
       "      <td>B.1.617.2.131</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>AY.131</td>\n",
       "      <td>AY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>B.1.1.384</td>\n",
       "      <td>Swiss Lineage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B.1.1.384</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>B.1</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        lineage                                        description  \\\n",
       "1177    B.1.418                                   Scottish lineage   \n",
       "1327    B.1.571                USA lineage, split out from B.1.369   \n",
       "1935    B.1.286  Withdrawn: Reassigned in the current tree. USA...   \n",
       "1622     AY.131  Alias of B.1.617.2.131, mainly reported from B...   \n",
       "488   B.1.1.384                                      Swiss Lineage   \n",
       "\n",
       "              alias predecessor         l0     l1   l2   l3 levels  \n",
       "1177            NaN         NaN    B.1.418    B.1    B  NaN      3  \n",
       "1327            NaN         NaN    B.1.571    B.1    B  NaN      3  \n",
       "1935            NaN         NaN    B.1.286    B.1    B  NaN      3  \n",
       "1622  B.1.617.2.131   B.1.617.2     AY.131     AY  NaN  NaN      2  \n",
       "488             NaN         NaN  B.1.1.384  B.1.1  B.1    B      4  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pango.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineages = pango['lineage'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern1 = re.compile(' [A-Z]{1,2}[.]\\d+ ', re.IGNORECASE)\n",
    "pattern2 = re.compile(' [A-Z]{1,2}[.]\\d+[.]\\d+ ', re.IGNORECASE)\n",
    "pattern3 = re.compile(' [A-Z]{1,2}[.]\\d+[.]\\d+[.]+\\d+ ', re.IGNORECASE)\n",
    "\n",
    "# add WHO lineage\n",
    "who_lineage = [' Alpha ', ' Beta ', ' Gamma ', ' Epsilon ',' Zeta ', ' Eta ', ' Theta  ',\\\n",
    "               ' Iota ', ' Kappa ', ' Lambda ', ' Mu ']\n",
    "pattern4 = re.compile(\"|\".join(who_lineage), re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add who to lineages\n",
    "lineages = np.append(lineages, who_lineage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove A B\n",
    "lineages = np.delete(lineages, np.where(lineages == 'A'))\n",
    "lineages = np.delete(lineages, np.where(lineages == 'B'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get CORD-19 Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "CACHE = Path(NEO4J_IMPORT +'/cache/cord19/2022-03-31/metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(CACHE, dtype='str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.fillna('', inplace=True)\n",
    "#convert datetime column to just date\n",
    "metadata['year'] = metadata['publish_time'].apply(lambda d: d[:4] if len(d) > 4 else '')\n",
    "metadata['date'] = metadata['publish_time'].apply(lambda d: dateutil.parser.parse(d) if len(d) > 0 else '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of papers 992921\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of papers\", metadata.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modify for knowledge graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>mag_id</th>\n",
       "      <th>who_covidence_id</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>pdf_json_files</th>\n",
       "      <th>pmc_json_files</th>\n",
       "      <th>url</th>\n",
       "      <th>s2_id</th>\n",
       "      <th>year</th>\n",
       "      <th>date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ug7v899j</td>\n",
       "      <td>d1aafb70c066a2068b02786f8929fd9c900897fb</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Clinical features of culture-proven Mycoplasma...</td>\n",
       "      <td>10.1186/1471-2334-1-6</td>\n",
       "      <td>PMC35282</td>\n",
       "      <td>11472636</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>OBJECTIVE: This retrospective chart review des...</td>\n",
       "      <td>2001-07-04</td>\n",
       "      <td>Madani, Tariq A; Al-Ghamdi, Aisha A</td>\n",
       "      <td>BMC Infect Dis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>document_parses/pdf_json/d1aafb70c066a2068b027...</td>\n",
       "      <td>document_parses/pmc_json/PMC35282.xml.json</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td></td>\n",
       "      <td>2001</td>\n",
       "      <td>2001-07-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>02tnwd4m</td>\n",
       "      <td>6b0567729c2143a66d737eb0a2f63f2dce2e5a7d</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Nitric oxide: a pro-inflammatory mediator in l...</td>\n",
       "      <td>10.1186/rr14</td>\n",
       "      <td>PMC59543</td>\n",
       "      <td>11667967</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Inflammatory diseases of the respiratory tract...</td>\n",
       "      <td>2000-08-15</td>\n",
       "      <td>Vliet, Albert van der; Eiserich, Jason P; Cros...</td>\n",
       "      <td>Respir Res</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>document_parses/pdf_json/6b0567729c2143a66d737...</td>\n",
       "      <td>document_parses/pmc_json/PMC59543.xml.json</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td></td>\n",
       "      <td>2000</td>\n",
       "      <td>2000-08-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha source_x  \\\n",
       "0  ug7v899j  d1aafb70c066a2068b02786f8929fd9c900897fb      PMC   \n",
       "1  02tnwd4m  6b0567729c2143a66d737eb0a2f63f2dce2e5a7d      PMC   \n",
       "\n",
       "                                               title                    doi  \\\n",
       "0  Clinical features of culture-proven Mycoplasma...  10.1186/1471-2334-1-6   \n",
       "1  Nitric oxide: a pro-inflammatory mediator in l...           10.1186/rr14   \n",
       "\n",
       "      pmcid pubmed_id license  \\\n",
       "0  PMC35282  11472636   no-cc   \n",
       "1  PMC59543  11667967   no-cc   \n",
       "\n",
       "                                            abstract publish_time  \\\n",
       "0  OBJECTIVE: This retrospective chart review des...   2001-07-04   \n",
       "1  Inflammatory diseases of the respiratory tract...   2000-08-15   \n",
       "\n",
       "                                             authors         journal mag_id  \\\n",
       "0                Madani, Tariq A; Al-Ghamdi, Aisha A  BMC Infect Dis          \n",
       "1  Vliet, Albert van der; Eiserich, Jason P; Cros...      Respir Res          \n",
       "\n",
       "  who_covidence_id arxiv_id  \\\n",
       "0                             \n",
       "1                             \n",
       "\n",
       "                                      pdf_json_files  \\\n",
       "0  document_parses/pdf_json/d1aafb70c066a2068b027...   \n",
       "1  document_parses/pdf_json/6b0567729c2143a66d737...   \n",
       "\n",
       "                               pmc_json_files  \\\n",
       "0  document_parses/pmc_json/PMC35282.xml.json   \n",
       "1  document_parses/pmc_json/PMC59543.xml.json   \n",
       "\n",
       "                                                 url s2_id  year       date  \n",
       "0  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...        2001 2001-07-04  \n",
       "1  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...        2000 2000-08-15  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This part can be removed, refer to \"hits\" part\n",
    "\"\"\"def kg_id(x):\n",
    "    if x['doi']: return x['doi']\n",
    "    elif x['pubmed_id']: return x['pubmed_id']\n",
    "    elif x['pmcid']: return x['pmcid']\n",
    "    elif x['url']: return x['url']\n",
    "    elif x['mag_id']: return x['mag_id']\n",
    "    elif x['who_covidence_id']: return x['who_covidence_id']\n",
    "    elif x['arxiv_id']: return x['arxiv_id']\n",
    "    elif x['s2_id']: return x['s2_id']\n",
    "    else: return cord_uid\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"metadata['id'] = metadata.apply(kg_id,axis=1)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#metadata['id'] = metadata['pmcid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Extract a list of PANGO lineages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove special characters to simply parsing for lineages in parenthesis, comma-separated lists, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['title'] = metadata['title'].replace('[()/,]', ' ', regex=True)\n",
    "metadata['abstract'] = metadata['abstract'].replace('[()/,]', ' ', regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Match PANGO patterns and check agains list of known lineages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern1 = re.compile(' [A-Z]{1,2}[.]\\d+ ', re.IGNORECASE)\n",
    "pattern2 = re.compile(' [A-Z]{1,2}[.]\\d+[.]\\d+ ', re.IGNORECASE)\n",
    "pattern3 = re.compile(' [A-Z]{1,2}[.]\\d+[.]\\d+[.]+\\d+ ', re.IGNORECASE)\n",
    "\n",
    "# add WHO lineage\n",
    "who_lineage = [' Alpha ', ' Beta ', ' Gamma ', ' Epsilon ',' Zeta ', ' Eta ', ' Theta  ',\\\n",
    "               ' Iota ', ' Kappa ', ' Lambda ', ' Mu ']\n",
    "pattern4 = re.compile(\"|\".join(who_lineage), re.IGNORECASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add who to lineages\n",
    "lineages = np.append(lineages, who_lineage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lineages(row):\n",
    "    text = ' ' + row.title + ' ' + row.abstract + ' '\n",
    "    lin = pattern1.findall(text) + pattern2.findall(text) + pattern3.findall(text)\n",
    "    u_lin = set()\n",
    "    \n",
    "    \n",
    "    for l in lin:\n",
    "        l = l.strip()\n",
    "        # check if lineage is valid (e.g., not a withdrawn lineage or false positive)\n",
    "        if l in lineages:\n",
    "            u_lin.add(l)\n",
    "            \n",
    "    return \";\".join(u_lin)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Run on whole dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['lineages'] = metadata.apply(get_lineages, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep only papers that map to PANGO lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = metadata[metadata['lineages'].str.len() > 0].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Assign CURIEs from [Identifiers.org](https://identifiers.org)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits['doi'] = hits['doi'].apply(lambda x: 'doi:' + x if len(x) > 0 else '')\n",
    "hits['pubmed_id'] = hits['pubmed_id'].apply(lambda x: 'pubmed:' + x if len(x) > 0 else '')\n",
    "hits['pmcid'] = hits['pmcid'].apply(lambda x: 'pmc:' + x if len(x) > 0 else '')\n",
    "hits['arxiv_id'] = hits['arxiv_id'].apply(lambda x: 'arxiv:' + x if len(x) > 0 else '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hits.sort_values(by=['publish_time'], ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matches 3205\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of matches\", hits.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_id(row):\n",
    "    \"\"\"Creates a unique id using the most commonly available id in priority order\"\"\"\n",
    "    if row.doi != '':\n",
    "        return row.doi\n",
    "    elif row.pubmed_id != '':\n",
    "        return row.pubmed_id\n",
    "    elif row.pmcid != '':\n",
    "        return row.pmcid\n",
    "    elif row.arxiv_id != '':\n",
    "        return row.arxiv_id\n",
    "    elif row.url != '':\n",
    "        return row.url\n",
    "    else:\n",
    "        # TODO deal with WHO papers here?\n",
    "        return ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits['id'] = hits.apply(create_id, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WHO documents seem to be copies of articles that are already present in the dataset and will be ignored for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits.query('id != \"\"', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of matches 3205\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of matches\", hits.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits.to_csv(NEO4J_IMPORT + \"01h-CORDLineages.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Knowledge graph \"Publications.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = hits[['id','title','doi','url','pubmed_id','pmcid','journal','year','title','abstract']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = ['id', 'name', 'doi', 'url', 'pubmedId', 'pmcId', 'journal', 'year',\n",
    "       'title', 'abstract']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32050"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only takes those have pmcid, otherwise it is impossible to scrape\n",
    "df = df[df.pmcId != '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('~/Desktop/Publication_11_13.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KG Generalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pub_mentions_lin(article_id, real_id):\n",
    "    body_text = download_article(article_id) # get body text\n",
    "    record = get_full_lineage(body_text) # extract lineages in text\n",
    "    [x.append(real_id) for x in record] # attach article id to lineage record\n",
    "    df = pd.DataFrame(record)\n",
    "    df.columns = ['lineage', 'string', 'ID']\n",
    "    df = df[['ID','string', 'lineage']]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def run_pipeline(N, pub):\n",
    "    results = []\n",
    "    for i in range(N):\n",
    "        article = pub.iloc[i]\n",
    "        article_id = article.pmcId.split(\":\")[1]\n",
    "        real_id = article.id\n",
    "        try:\n",
    "            results.append(pub_mentions_lin(article_id, real_id))\n",
    "        except urllib.error.HTTPError as exc:\n",
    "            time.sleep(5) # wait 5 seconds and then make http request again\n",
    "            continue\n",
    "    return pd.concat(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('Publication.csv')\n",
    "pub = df.copy()\n",
    "N = pub.size\n",
    "run_pipeline(N, pub)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Fulltext Regrex\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get articles ids for specific lineage\n",
    "def get_ids(lineage):\n",
    "    url = requests.get(f'https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=(%22{lineage}%22%20AND%20(%22SARS-CoV-2%22%20OR%20%22COVID-19%22)%20AND%20(%22lineage%22%20OR%20%22lineages%22%20OR%20%22strain%22%20OR%20%22strains%22%20OR%20%22variants%22%20OR%20%22variants%22))%20AND%20(FIRST_PDATE:%5b2020-01-01%20)%20AND%20HAS_FT:y%20AND%20%20sort_date:y&resultType=idlist&pageSize=1000&format=json&cursorMark=*')\n",
    "    text = url.text\n",
    "    results = json.loads(text)['resultList']['result']\n",
    "    ids = list(map(lambda x: x['fullTextIdList']['fullTextId'][0], results))\n",
    "    return ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download articles in XML and return body paragraph\n",
    "def download_article(article_id):\n",
    "    url = f'https://www.ebi.ac.uk/europepmc/webservices/rest/{article_id}/fullTextXML'\n",
    "    print(url)\n",
    "    xmldoc = parse(urlopen(url))\n",
    "    \n",
    "    # get full text\n",
    "    root = xmldoc.getroot()\n",
    "    text = root.findall('.//p')\n",
    "\n",
    "    # put body paragraphs together\n",
    "    ptext = \"\"\n",
    "    for p in text:\n",
    "        ptext += ''.join([x for x in p.itertext()]) + '.\\n' + '\\n'\n",
    "    return ptext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get lineage for full texts\n",
    "def get_full_lineage(ptext):\n",
    "    # tokenize texts into sentences\n",
    "    p_sentence = nltk.tokenize.sent_tokenize(ptext)\n",
    "    \n",
    "    # record lineages\n",
    "    linset = set()\n",
    "    pair = []\n",
    "    for s in p_sentence:\n",
    "        s1 = re.subn('[()/,]', ' ', s)[0] # remove special chars\n",
    "        lin = set(pattern1.findall(s1) + pattern2.findall(s1) + pattern3.findall(s1) + pattern4.findall(s1))\n",
    "\n",
    "        if lin: \n",
    "            for l in lin:\n",
    "                # valid lineage and not recorded\n",
    "                l = l.strip()\n",
    "                l = l.capitalize()\n",
    "                if (l in lineages) and (l not in linset): \n",
    "                    linset.add(l)\n",
    "                    pair.append([l, s])\n",
    "                else: continue\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "    ptext = re.subn('[()/,]', ' ', ptext)[0] # remove special chars\n",
    "    lin = pattern1.findall(ptext) + pattern2.findall(ptext) + pattern3.findall(ptext)\n",
    "    lin_set = set(lin)\n",
    "    \n",
    "    record = []\n",
    "    if lin_set:\n",
    "        for l in lin_set:\n",
    "            \n",
    "            sen = re.search(r\"\\.?([^\\.]*{}[^\\.]*)\".format(l), ptext).group()\n",
    "            record.append([l, sen])\n",
    "    \"\"\"\n",
    "    return pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wrap up function take lineage ids as input and output dataframe\n",
    "def extract_full(ids):\n",
    "    full_regrex = []\n",
    "    \n",
    "    if not ids:  \n",
    "        return None\n",
    "    for i in ids:\n",
    "        try:\n",
    "            body_text = download_article(i) # get body text\n",
    "            record = get_full_lineage(body_text) # extract lineages in text\n",
    "            [x.append(i) for x in record] # attach article id to lineage record\n",
    "            full_regrex.append(pd.DataFrame(record))\n",
    "        except urllib.error.HTTPError as exc:\n",
    "            time.sleep(5) # wait 10 seconds and then make http request again\n",
    "            continue\n",
    "    df_fulltext = pd.concat(full_regrex)\n",
    "    df_fulltext.columns = ['lineage', 'string', 'ID']\n",
    "    return df_fulltext\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### manual check possible false postives\n",
    "#### B.1.1.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b117 = pd.read_csv('B_1_1_7.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b117 = b117.drop(['Unnamed: 0'],axis = 1)\n",
    "b117.columns = ['lineage', 'string contains lineage', 'ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. same sentence with many lineages are counted as positive, we remove those \n",
    "b117_sub = b117[~ b117.duplicated('string contains lineage',keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. extract articles with only one lineage, which are possibly FP\n",
    "IDs = b117_sub.groupby('ID').lineage.count().loc[lambda p : p == 1].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b117_sub = b117_sub.set_index('ID').loc[IDs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manual check\n",
    "#print(b117_sub['string contains lineage'].str.cat(sep = '\\n '))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### P.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = pd.read_csv(\"P_1.csv\")\n",
    "p1_sub = p1[~ p1.duplicated('string contains lineage',keep=False)]\n",
    "IDs_1 = p1_sub.groupby('ID').lineage.count().loc[lambda p : p == 1].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1_sub = p1_sub.set_index('ID').loc[IDs_1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1_sub[p1_sub['string contains lineage'].str.contains('The nucleic acid')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "def iter_lineage(l_):\n",
    "    \n",
    "    start = time.time()\n",
    "    print(f'start {l_}')\n",
    "    #get ids\n",
    "    ids = get_ids(l_)\n",
    "    try:\n",
    "        # if lineage processed before, run on new ids\n",
    "        path = Path(f'{l_}_df.csv')\n",
    "        if path.is_file():\n",
    "            print('exist')\n",
    "            df = pd.read_csv(path)\n",
    "            id_ran = set(df.ID)\n",
    "            id_to_run = set(ids) - id_ran\n",
    "            print(f'new id {id_to_run}')\n",
    "            df_new = extract_full(id_to_run) # merge with original df\n",
    "            ddf = pd.concat([df, df_new], axis = 0)\n",
    "            print('done merge')\n",
    "\n",
    "        else: # otherwise run on all ids\n",
    "            id_to_run = ids\n",
    "            ddf = extract_full(id_to_run)\n",
    "\n",
    "        ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                    header = ['lineage', 'string', 'ID'])\n",
    "        end = time.time()\n",
    "        print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "    except:\n",
    "        id_to_run = ids\n",
    "        ddf = extract_full(id_to_run)\n",
    "        ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                    header = ['lineage', 'string', 'ID'])\n",
    "        end = time.time()\n",
    "        print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i1 = get_ids('A.1')\n",
    "i2 = get_ids('A.2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1 = extract_full(i1[:10])\n",
    "d2 = extract_full(i2[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i1[:5],i2[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A.1', 'A.2', 'A.2.2', 'A.2.3', 'A.2.4', 'A.2.5', 'A.2.5.1',\n",
       "       'A.2.5.2', 'A.2.5.3', 'A.3'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_10 = lineages[:10]\n",
    "l_10\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parallel running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask.distributed import Client, progress"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "    <div style=\"width: 24px; height: 24px; background-color: #e1e1e1; border: 3px solid #9D9D9D; border-radius: 5px; position: absolute;\"> </div>\n",
       "    <div style=\"margin-left: 48px;\">\n",
       "        <h3 style=\"margin-bottom: 0px;\">Client</h3>\n",
       "        <p style=\"color: #9D9D9D; margin-bottom: 0px;\">Client-9d0cab62-04d9-11ed-9ebc-8c8590c1bce1</p>\n",
       "        <table style=\"width: 100%; text-align: left;\">\n",
       "\n",
       "        <tr>\n",
       "        \n",
       "            <td style=\"text-align: left;\"><strong>Connection method:</strong> Cluster object</td>\n",
       "            <td style=\"text-align: left;\"><strong>Cluster type:</strong> distributed.LocalCluster</td>\n",
       "        \n",
       "        </tr>\n",
       "\n",
       "        \n",
       "            <tr>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:8787/status\" target=\"_blank\">http://127.0.0.1:8787/status</a>\n",
       "                </td>\n",
       "                <td style=\"text-align: left;\"></td>\n",
       "            </tr>\n",
       "        \n",
       "\n",
       "        </table>\n",
       "\n",
       "        \n",
       "            <details>\n",
       "            <summary style=\"margin-bottom: 20px;\"><h3 style=\"display: inline;\">Cluster Info</h3></summary>\n",
       "            <div class=\"jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output\">\n",
       "    <div style=\"width: 24px; height: 24px; background-color: #e1e1e1; border: 3px solid #9D9D9D; border-radius: 5px; position: absolute;\">\n",
       "    </div>\n",
       "    <div style=\"margin-left: 48px;\">\n",
       "        <h3 style=\"margin-bottom: 0px; margin-top: 0px;\">LocalCluster</h3>\n",
       "        <p style=\"color: #9D9D9D; margin-bottom: 0px;\">c9e29805</p>\n",
       "        <table style=\"width: 100%; text-align: left;\">\n",
       "            <tr>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Dashboard:</strong> <a href=\"http://127.0.0.1:8787/status\" target=\"_blank\">http://127.0.0.1:8787/status</a>\n",
       "                </td>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Workers:</strong> 4\n",
       "                </td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Total threads:</strong> 4\n",
       "                </td>\n",
       "                <td style=\"text-align: left;\">\n",
       "                    <strong>Total memory:</strong> 16.00 GiB\n",
       "                </td>\n",
       "            </tr>\n",
       "            \n",
       "            <tr>\n",
       "    <td style=\"text-align: left;\"><strong>Status:</strong> running</td>\n",
       "    <td style=\"text-align: left;\"><strong>Using processes:</strong> True</td>\n",
       "</tr>\n",
       "\n",
       "            \n",
       "        </table>\n",
       "\n",
       "        <details>\n",
       "            <summary style=\"margin-bottom: 20px;\">\n",
       "                <h3 style=\"display: inline;\">Scheduler Info</h3>\n",
       "            </summary>\n",
       "\n",
       "            <div style=\"\">\n",
       "    <div>\n",
       "        <div style=\"width: 24px; height: 24px; background-color: #FFF7E5; border: 3px solid #FF6132; border-radius: 5px; position: absolute;\"> </div>\n",
       "        <div style=\"margin-left: 48px;\">\n",
       "            <h3 style=\"margin-bottom: 0px;\">Scheduler</h3>\n",
       "            <p style=\"color: #9D9D9D; margin-bottom: 0px;\">Scheduler-346382f3-8cc9-4e04-9749-cd8e201ba4aa</p>\n",
       "            <table style=\"width: 100%; text-align: left;\">\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Comm:</strong> tcp://127.0.0.1:64218\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Workers:</strong> 4\n",
       "                    </td>\n",
       "                </tr>\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Dashboard:</strong> <a href=\"http://127.0.0.1:8787/status\" target=\"_blank\">http://127.0.0.1:8787/status</a>\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Total threads:</strong> 4\n",
       "                    </td>\n",
       "                </tr>\n",
       "                <tr>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Started:</strong> Just now\n",
       "                    </td>\n",
       "                    <td style=\"text-align: left;\">\n",
       "                        <strong>Total memory:</strong> 16.00 GiB\n",
       "                    </td>\n",
       "                </tr>\n",
       "            </table>\n",
       "        </div>\n",
       "    </div>\n",
       "\n",
       "    <details style=\"margin-left: 48px;\">\n",
       "        <summary style=\"margin-bottom: 20px;\">\n",
       "            <h3 style=\"display: inline;\">Workers</h3>\n",
       "        </summary>\n",
       "\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 0</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:64229\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 1\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:64230/status\" target=\"_blank\">http://127.0.0.1:64230/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 4.00 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:64221\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /Users/lyt/Desktop/COVID-19/covid-19-community/notebooks/dataprep/dask-worker-space/worker-a7k7unbr\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 1</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:64238\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 1\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:64239/status\" target=\"_blank\">http://127.0.0.1:64239/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 4.00 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:64224\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /Users/lyt/Desktop/COVID-19/covid-19-community/notebooks/dataprep/dask-worker-space/worker-oaym1bka\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 2</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:64232\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 1\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:64233/status\" target=\"_blank\">http://127.0.0.1:64233/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 4.00 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:64222\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /Users/lyt/Desktop/COVID-19/covid-19-community/notebooks/dataprep/dask-worker-space/worker-t3s4zw8r\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "        <div style=\"margin-bottom: 20px;\">\n",
       "            <div style=\"width: 24px; height: 24px; background-color: #DBF5FF; border: 3px solid #4CC9FF; border-radius: 5px; position: absolute;\"> </div>\n",
       "            <div style=\"margin-left: 48px;\">\n",
       "            <details>\n",
       "                <summary>\n",
       "                    <h4 style=\"margin-bottom: 0px; display: inline;\">Worker: 3</h4>\n",
       "                </summary>\n",
       "                <table style=\"width: 100%; text-align: left;\">\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Comm: </strong> tcp://127.0.0.1:64235\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Total threads: </strong> 1\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Dashboard: </strong> <a href=\"http://127.0.0.1:64236/status\" target=\"_blank\">http://127.0.0.1:64236/status</a>\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Memory: </strong> 4.00 GiB\n",
       "                        </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td style=\"text-align: left;\">\n",
       "                            <strong>Nanny: </strong> tcp://127.0.0.1:64223\n",
       "                        </td>\n",
       "                        <td style=\"text-align: left;\"></td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <td colspan=\"2\" style=\"text-align: left;\">\n",
       "                            <strong>Local directory: </strong> /Users/lyt/Desktop/COVID-19/covid-19-community/notebooks/dataprep/dask-worker-space/worker-piy54tdf\n",
       "                        </td>\n",
       "                    </tr>\n",
       "\n",
       "                    \n",
       "\n",
       "                    \n",
       "\n",
       "                </table>\n",
       "            </details>\n",
       "            </div>\n",
       "        </div>\n",
       "        \n",
       "\n",
       "    </details>\n",
       "</div>\n",
       "\n",
       "        </details>\n",
       "    </div>\n",
       "</div>\n",
       "            </details>\n",
       "        \n",
       "\n",
       "    </div>\n",
       "</div>"
      ],
      "text/plain": [
       "<Client: 'tcp://127.0.0.1:64218' processes=4 threads=4, memory=16.00 GiB>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "client = Client(n_workers=4, threads_per_worker=1, memory_limit=\"4 GiB\")\n",
    "client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask.bag as db\n",
    "\n",
    "b = db.from_sequence(l_10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/3496112380.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0miter_lineage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mans\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Total time:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m         \"\"\"\n\u001b[0;32m--> 290\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mpostcomputes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dask_postcompute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 573\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    574\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)\u001b[0m\n\u001b[1;32m   2992\u001b[0m                     \u001b[0mshould_rejoin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2993\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2994\u001b[0;31m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2995\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2996\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfutures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mgather\u001b[0;34m(self, futures, errors, direct, asynchronous)\u001b[0m\n\u001b[1;32m   2150\u001b[0m                 \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2151\u001b[0m                 \u001b[0mlocal_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlocal_worker\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2152\u001b[0;31m                 \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2153\u001b[0m             )\n\u001b[1;32m   2154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    308\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    309\u001b[0m             return sync(\n\u001b[0;32m--> 310\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    311\u001b[0m             )\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    370\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_set\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 372\u001b[0;31m             \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(timeout)\u001b[0m\n\u001b[1;32m    359\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0mloop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcancel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    550\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 552\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    553\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    554\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    298\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "result = b.map(lambda l: iter_lineage(l))\n",
    "ans = result.compute()\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = db.from_sequence(a,npartitions = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normal loop running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start A.2.5.1\n",
      "{\"ver\n",
      "[           0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647,            0                                                  1           2\n",
      "0  B.1.617.2  The dominant lineage in most of the 28 states ...  PMC9115369\n",
      "1     B.1.36  However, the states Madhya Pradesh, Jammu & Ka...  PMC9115369\n",
      "2    B.1.1.7  However, the states Madhya Pradesh, Jammu & Ka...  PMC9115369\n",
      "3  B.1.36.29  B.1.617.2‐Delta variant was the major dominati...  PMC9115369\n",
      "4        B.1  B.1.617.2‐Delta variant was the major dominati...  PMC9115369\n",
      "5      B.1.1  B.1.617.2‐Delta variant was the major dominati...  PMC9115369\n",
      "6  B.1.1.216  7%), B.1.1.216 (1.8%), B.1.1.306 (1.8%), AY12 ...  PMC9115369\n",
      "7  B.1.1.306  7%), B.1.1.216 (1.8%), B.1.1.306 (1.8%), AY12 ...  PMC9115369\n",
      "8  B.1.617.1  It has been reported that B.1.617.1 and B.1.61...  PMC9115369,              0                                                  1           2\n",
      "0      B.1.427  The massive emergence of L452R variants was fi...  PMC8525575\n",
      "1      B.1.232  By PCR amplification and Sanger sequencing of ...  PMC8525575\n",
      "2     B.1.1.28  Because of the important functional and antige...  PMC8525575\n",
      "3      B.1.1.7  Because of the important functional and antige...  PMC8525575\n",
      "4      B.1.351  Because of the important functional and antige...  PMC8525575\n",
      "5    B.1.617.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "6       B.1.36  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "7   B.1.177.83  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "8    B.1.1.459  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "9          P.4  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "10   B.1.617.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "11   B.1.1.263  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "12      C.36.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "13        A.21  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "14   B.1.526.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "15      C.36.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "16     A.2.5.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "17     B.1.429  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "18   B.1.617.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "19        C.37  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "20        C.16  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "21     A.2.5.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "22     B.1.466  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "23        A.27  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575]\n",
      "done with A.2.5.1, time duration --- seconds --- 30.76284098625183 \n",
      "\n",
      "start A.2.5.2\n",
      "{\"ver\n",
      "{\"ver\n",
      "{\"ver\n",
      "{\"ver\n",
      "{\"ver\n",
      "start A.2.2\n",
      "exist\n",
      "{\"ver\n",
      "{'PMC9205280', 'PMC7650373', 'PMC8251645', 'PMC7741497'}\n",
      "start A.2.3\n",
      "exist\n",
      "{'PMC7350918', 'PMC8390340'}\n",
      "hello\n",
      "start A.2.2\n",
      "{\"ver\n",
      "done with A.2.3, time duration --- seconds --- 64.69221186637878 \n",
      "\n",
      "done with A.2.2, time duration --- seconds --- 367.4889380931854 \n",
      "\n",
      "[]\n",
      "{\"ver\n",
      "{\"ver\n",
      "{\"ver\n",
      "start ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3']\n",
      "{\"ver\n",
      "[           0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647]\n",
      "done with ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3'], time duration --- seconds --- 6.234965085983276 \n",
      "\n",
      "start ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3']\n",
      "{\"ver\n",
      "exist\n",
      "new id set()\n",
      "done merge\n",
      "done with ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3'], time duration --- seconds --- 1.7190001010894775 \n",
      "\n",
      "start ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3']\n",
      "{\"ver\n",
      "exist\n",
      "new id set()\n",
      "done merge\n",
      "done with ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3'], time duration --- seconds --- 1.6131610870361328 \n",
      "\n",
      "start ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3']\n",
      "{\"ver\n",
      "exist\n",
      "new id set()\n",
      "done merge\n",
      "done with ['A.1' 'A.2' 'A.2.2' 'A.2.3' 'A.2.4' 'A.2.5' 'A.2.5.1' 'A.2.5.2' 'A.2.5.3'\n",
      " 'A.3'], time duration --- seconds --- 3.3464338779449463 \n",
      "\n",
      "start A.2.4\n",
      "{\"ver\n",
      "[Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],            0                                                  1           2\n",
      "0    B.1.351  As of November 25th 2021, four SARS-CoV − 2 va...  PMC9034743\n",
      "1    B.1.1.7  As of November 25th 2021, four SARS-CoV − 2 va...  PMC9034743\n",
      "2  B.1.617.2  As of November 25th 2021, four SARS-CoV − 2 va...  PMC9034743\n",
      "3        P.1  As of November 25th 2021, four SARS-CoV − 2 va...  PMC9034743,            0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647, Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],            0                                                  1           2\n",
      "0  B.1.1.529  On 26 November 2021, a new variant named Omicr...  PMC8876558\n",
      "1    B.1.351  Until late 2021, four variants of concern (VOC...  PMC8876558\n",
      "2    B.1.1.7  Until late 2021, four variants of concern (VOC...  PMC8876558\n",
      "3  B.1.617.2  Until late 2021, four variants of concern (VOC...  PMC8876558\n",
      "4        P.1  Until late 2021, four variants of concern (VOC...  PMC8876558\n",
      "5      B.1.1  Moreover, the B.1.1 variants and human coronav...  PMC8876558\n",
      "6    B.1.526  High hopes come from a recently found RBD-spec...  PMC8876558\n",
      "7    B.1.427  High hopes come from a recently found RBD-spec...  PMC8876558\n",
      "8  B.1.640.2  In fact, while the world is grappling with the...  PMC8876558,            0                                                  1          2\n",
      "0  B.1.1.529  A newly emerged SARS-CoV-2 variant B.1.1.529 h...  PPR435130\n",
      "1        B.1  4)..\\n\\nOur analysis of the SARS-CoV-2 genomic...  PPR435130\n",
      "2    B.1.621  P681H is characteristically present in multipl...  PPR435130\n",
      "3        Q.1  P681H is characteristically present in multipl...  PPR435130\n",
      "4    B.1.1.7  P681H is characteristically present in multipl...  PPR435130\n",
      "5        P.1  P681H is characteristically present in multipl...  PPR435130,            0                                                  1           2\n",
      "0    B.1.234  Unsupervised learning revealed eight VRV-haplo...  PMC8777887\n",
      "1    B.1.1.7  E484, an AA-sub in the B.1.1.7, P.1, and B.1.3...  PMC8777887\n",
      "2    B.1.351  E484, an AA-sub in the B.1.1.7, P.1, and B.1.3...  PMC8777887\n",
      "3        P.1  E484, an AA-sub in the B.1.1.7, P.1, and B.1.3...  PMC8777887\n",
      "4    B.1.429  We performed homology modeling on two potentia...  PMC8777887\n",
      "5  B.1.1.222  Among these pressing VRVs, the haplotype (T478...  PMC8777887,             0                                                  1           2\n",
      "0   B.1.617.2  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "1   B.1.617.1  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "2     B.1.617  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "3        C.37  In comparison, similar modest reductions were ...  PMC8707919\n",
      "4     B.1.526  In comparison, similar modest reductions were ...  PMC8707919\n",
      "5     B.1.351  In comparison, similar modest reductions were ...  PMC8707919\n",
      "6         R.1  In comparison, similar modest reductions were ...  PMC8707919\n",
      "7         P.1  In comparison, similar modest reductions were ...  PMC8707919\n",
      "8         B.1  One of the earliest variants that is highly in...  PMC8707919\n",
      "9     B.1.429  Subsequent genomic surveillance has led to the...  PMC8707919\n",
      "10    B.1.427  The B.1.427/B.1.429 and B.1.617 lineage varian...  PMC8707919\n",
      "11        P.3  The E484K substitution found in B.1.351, B.1.5...  PMC8707919\n",
      "12    B.1.1.7  These findings extend a prior study reporting ...  PMC8707919\n",
      "13      B.1.2  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "14      C.1.2  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "15    B.1.525  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "16  B.1.617.3  Sequences corresponding to the S1/S2 region of...  PMC8707919,            0                                                  1          2\n",
      "0        P.1  Genomic surveillance of SC2 variants has been ...  PPR428671\n",
      "1      A.2.5  Here, we describe the introduction of SC2 line...  PPR428671\n",
      "2      A.2.4  In January 2021, BPHL detected the first cases...  PPR428671\n",
      "3        B.1  The A.2.5 lineage in Florida became of particu...  PPR428671\n",
      "4    B.1.427  A.2.5 also contains spike L452R, a CDC-defined...  PPR428671\n",
      "5  B.1.617.1  Allele L452R has independently emerged in seve...  PPR428671\n",
      "6    B.1.526  Allele L452R has independently emerged in seve...  PPR428671\n",
      "7  B.1.617.2  Allele L452R has independently emerged in seve...  PPR428671\n",
      "8       C.36  Allele L452R has independently emerged in seve...  PPR428671\n",
      "9    B.1.1.7  Additional mutations causing increased transmi...  PPR428671,            0                                                  1          2\n",
      "0  B.1.617.2  Epidemiological data demonstrate that SARS-CoV...  PPR416637\n",
      "1    B.1.1.7  Epidemiological data demonstrate that SARS-CoV...  PPR416637\n",
      "2        B.1  B.1.1.7 virus isolates displayed inferior or e...  PPR416637\n",
      "3    B.1.351  In accordance with reports by others (Bates et...  PPR416637\n",
      "4        P.2  Briefly, cells were expanded using the conditi...  PPR416637\n",
      "5        P.3  Briefly, cells were expanded using the conditi...  PPR416637,            0                                                  1           2\n",
      "0    B.1.351  The new variants are accumulating mutations in...  PMC8535385\n",
      "1    B.1.1.7  The new variants are accumulating mutations in...  PMC8535385\n",
      "2  B.1.617.2  The new variants are accumulating mutations in...  PMC8535385\n",
      "3        P.1  The new variants are accumulating mutations in...  PMC8535385,             0                                                  1          2\n",
      "0     B.1.234  Unsupervised learning revealed eight VRV-haplo...  PPR357574\n",
      "1     B.1.427  The CDC is currently monitoring and characteri...  PPR357574\n",
      "2   B.1.617.1  The CDC is currently monitoring and characteri...  PPR357574\n",
      "3     B.1.429  The CDC is currently monitoring and characteri...  PPR357574\n",
      "4     B.1.526  The CDC is currently monitoring and characteri...  PPR357574\n",
      "5     B.1.351  The CDC is currently monitoring and characteri...  PPR357574\n",
      "6         P.2  The CDC is currently monitoring and characteri...  PPR357574\n",
      "7     B.1.1.7  The CDC is currently monitoring and characteri...  PPR357574\n",
      "8   B.1.617.3  The CDC is currently monitoring and characteri...  PPR357574\n",
      "9   B.1.617.2  The CDC is currently monitoring and characteri...  PPR357574\n",
      "10  B.1.526.1  The CDC is currently monitoring and characteri...  PPR357574\n",
      "11    B.1.617  The CDC is currently monitoring and characteri...  PPR357574\n",
      "12        P.1  The CDC is currently monitoring and characteri...  PPR357574\n",
      "13    B.1.525  The CDC is currently monitoring and characteri...  PPR357574\n",
      "14  B.1.1.222  Among these pressing VRVs, the haplotype (T478...  PPR357574, Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],           0                                                  1           2\n",
      "0   B.1.1.7  Among these mutations are amino acid changes N...  PMC8139174\n",
      "1   B.1.351  Among these mutations are amino acid changes N...  PMC8139174\n",
      "2       P.2  In December 2020, new variants of concern have...  PMC8139174\n",
      "3  B.1.1.74  In December 2020, new variants of concern have...  PMC8139174\n",
      "4       P.1  In December 2020, new variants of concern have...  PMC8139174,             0                                                  1           2\n",
      "0     B.1.1.7  Mutations of interest (for example those found...  PMC8290271\n",
      "1         B.1  Notably, the major clade containing D614G (Pan...  PMC8290271\n",
      "2        A.19  On the other hand, there are several examples ...  PMC8290271\n",
      "3       A.2.4  On the other hand, there are several examples ...  PMC8290271\n",
      "4        A.23  One exception is a lineage of viruses identifi...  PMC8290271\n",
      "5     B.1.318  A recurrent NTD deletion in the 140–145 region...  PMC8290271\n",
      "6         P.3  A recurrent NTD deletion in the 140–145 region...  PMC8290271\n",
      "7     B.1.525  A recurrent NTD deletion in the 140–145 region...  PMC8290271\n",
      "8         P.2  In recent months, several independent lineages...  PMC8290271\n",
      "9     B.1.351  In recent months, several independent lineages...  PMC8290271\n",
      "10        P.1  In recent months, several independent lineages...  PMC8290271\n",
      "11  B.1.1.318  The rapidly expanding list includes the B.1.52...  PMC8290271\n",
      "12    B.1.526  The rapidly expanding list includes the B.1.52...  PMC8290271\n",
      "13     A.23.1  The rapidly expanding list includes the B.1.52...  PMC8290271\n",
      "14       A.27  The rapidly expanding list includes the B.1.52...  PMC8290271\n",
      "15    B.1.258  The predominant lineage containing N439K (B.1....  PMC8290271\n",
      "16    B.1.429  In July 2020 a pair of sister lineages of SARS...  PMC8290271\n",
      "17    B.1.427  In July 2020 a pair of sister lineages of SARS...  PMC8290271\n",
      "18    B.1.177  There is speculation the main A222V-containing...  PMC8290271,             0                                                  1           2\n",
      "0     B.1.1.7  In this review we focus on the main variants o...  PMC8250244\n",
      "1     B.1.351  In this review we focus on the main variants o...  PMC8250244\n",
      "2         P.1  In this review we focus on the main variants o...  PMC8250244\n",
      "3       C.1.1  In April 2020, a preliminary work by the Londo...  PMC8250244\n",
      "4         B.1  29\\n showed positive selection of D614G, S477N...  PMC8250244\n",
      "5    B.1.1.33  Comparison of B.1.1.7, B1.1.28‐derived clades,...  PMC8250244\n",
      "6   B.1.1.207  66\\n, \\n67\\n, \\n68\\n It has been found both in...  PMC8250244\n",
      "7     B.1.346  36\\n In summary, it occurs in lineages B.1.375...  PMC8250244\n",
      "8     B.1.160  36\\n In summary, it occurs in lineages B.1.375...  PMC8250244\n",
      "9     B.1.177  36\\n In summary, it occurs in lineages B.1.375...  PMC8250244\n",
      "10  B.1.1.298  36\\n In summary, it occurs in lineages B.1.375...  PMC8250244\n",
      "11     A.23.1  It is found in the A.23.1 lineage from Uganda ...  PMC8250244\n",
      "12    B.1.232  78\\n L452R is the is the only Spike mutation f...  PMC8250244\n",
      "13    B.1.429  78\\n L452R is the is the only Spike mutation f...  PMC8250244\n",
      "14  B.1.1.130  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "15     B.1.74  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "16       C.16  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "17       A.21  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "18      A.2.4  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "19   B.1.1.10  99\\n L452R is also found in A.21, A.2.4, B.1.1...  PMC8250244\n",
      "20    B.1.525  V2 lineage from South Africa together with N50...  PMC8250244\n",
      "21    B.1.526  V2 lineage from South Africa together with N50...  PMC8250244\n",
      "22   B.1.1.28  The original B.1.1.28 lineage emerged in Brazi...  PMC8250244\n",
      "23  B.1.1.248  P.1 (also improperly termed B.1.1.28.1 or B.1....  PMC8250244\n",
      "24     B.1.91  172\\n Simultaneous infection by B.1.1.248 (eit...  PMC8250244\n",
      "25    B.1.258  Detecting the ΔHV69/70 deletion alone is not d...  PMC8250244\n",
      "26    B.1.375  Detecting the ΔHV69/70 deletion alone is not d...  PMC8250244, Empty DataFrame\n",
      "Columns: []\n",
      "Index: []]\n",
      "done with A.2.4, time duration --- seconds --- 110.03250908851624 \n",
      "\n",
      "start A.2.5\n",
      "{\"ver\n",
      "exist\n",
      "new id {'PMC7350918', 'PMC8021487', 'PMC7741497', 'PMC8189619', 'PMC8408311', 'PMC7650373', 'PMC9189962'}\n",
      "[Empty DataFrame\n",
      "Columns: []\n",
      "Index: [], Empty DataFrame\n",
      "Columns: []\n",
      "Index: [], Empty DataFrame\n",
      "Columns: []\n",
      "Index: [], Empty DataFrame\n",
      "Columns: []\n",
      "Index: [], Empty DataFrame\n",
      "Columns: []\n",
      "Index: []]\n",
      "[Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],            0                                                  1           2\n",
      "0    B.1.351  Since the original strain of severe acute resp...  PMC9119066\n",
      "1    B.1.1.7  Since the original strain of severe acute resp...  PMC9119066\n",
      "2  B.1.1.529  Since the original strain of severe acute resp...  PMC9119066\n",
      "3  B.1.617.2  Since the original strain of severe acute resp...  PMC9119066\n",
      "4        P.1  Since the original strain of severe acute resp...  PMC9119066\n",
      "5        Q.3  Among the 732 clinical samples that were teste...  PMC9119066\n",
      "6    B.1.621  By WGS, these samples were variant of interest...  PMC9119066,            0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647,             0                                                  1           2\n",
      "0     B.1.1.7  The three largest cities in Colombia: Bogotá, ...  PMC9008484\n",
      "1     B.1.621  The three largest cities in Colombia: Bogotá, ...  PMC9008484\n",
      "2   B.1.617.2  The three largest cities in Colombia: Bogotá, ...  PMC9008484\n",
      "3         P.1  The three largest cities in Colombia: Bogotá, ...  PMC9008484\n",
      "4       B.1.1  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "5        C.37  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "6         B.1  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "7       A.2.5  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "8     B.1.625  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "9   B.1.621.1  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "10     C.37.1  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "11  B.1.1.348  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "12  B.1.1.487  The figure shows the number of cases for SARS-...  PMC9008484\n",
      "13     P.1.10  The figure shows the number of cases for SARS-...  PMC9008484,            0                                                  1           2\n",
      "0  B.1.617.2  We also look at the effectiveness of approved ...  PMC8902153\n",
      "1    B.1.351  We also look at the effectiveness of approved ...  PMC8902153\n",
      "2        P.1  We also look at the effectiveness of approved ...  PMC8902153\n",
      "3    B.1.1.7  The D614G substitution is found in most of the...  PMC8902153\n",
      "4  B.1.1.529  The D614G substitution is found in most of the...  PMC8902153\n",
      "5    B.1.429  In addition, some of the notable substitutions...  PMC8902153\n",
      "6    B.1.617  In addition, some of the notable substitutions...  PMC8902153\n",
      "7      B.1.1  Fitness-enhancing mutations have been reported...  PMC8902153\n",
      "8  B.1.617.1  However, in-vitro neutralization assays show t...  PMC8902153\n",
      "9    B.1.427  According to the CDC, US, the P.1 variant seem...  PMC8902153,            0                                                  1          2\n",
      "0    B.1.1.7  Since the original strain of SARS-CoV-2 virus ...  PPR452207\n",
      "1    B.1.351  *), Beta (B.1.351), Gamma (P.1 and sublineages...  PPR452207\n",
      "2  B.1.617.2  *), Beta (B.1.351), Gamma (P.1 and sublineages...  PPR452207\n",
      "3        P.1  *), Beta (B.1.351), Gamma (P.1 and sublineages...  PPR452207\n",
      "4  B.1.1.529               *) and Omicron (B.1.1.529 and BA.*).  PPR452207\n",
      "5        Q.3  Among the 732 clinical samples that were teste...  PPR452207\n",
      "6    B.1.621  By WGS, these samples were variant of interest...  PPR452207\n",
      "7  B.1.1.519                 *), and Omicron (B.1.1.519 and BA.  PPR452207,            0                                                  1           2\n",
      "0       C.37  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "1    B.1.351  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "2    B.1.1.7  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "3  B.1.1.529  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "4    B.1.621  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "5  B.1.617.2  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "6        P.1  As of December 2021, the World Health Organiza...  PMC8942852\n",
      "7      B.1.1  On 24 November 2021, the B.1.1.529 lineage, a ...  PMC8942852,             0                                                  1           2\n",
      "0     B.1.351  A total of 411 samples belonging to the Alpha ...  PMC8769358\n",
      "1     B.1.1.7  A total of 411 samples belonging to the Alpha ...  PMC8769358\n",
      "2   B.1.617.2  A total of 411 samples belonging to the Alpha ...  PMC8769358\n",
      "3         P.1  A total of 411 samples belonging to the Alpha ...  PMC8769358\n",
      "4     B.1.427  A total of 411 samples belonging to the Alpha ...  PMC8769358\n",
      "5     B.1.369  The second largest clade identified by Nextcla...  PMC8769358\n",
      "6         B.1  The second largest clade identified by Nextcla...  PMC8769358\n",
      "7       B.1.2  Pangolin classified most of these as the US st...  PMC8769358\n",
      "8     B.1.526  A clear trend can be observed with the average...  PMC8769358\n",
      "9       B.1.1  While this mutation was observed in the B.1.1....  PMC8769358\n",
      "10    B.1.621  While this mutation was observed in the B.1.1....  PMC8769358\n",
      "11    B.1.324  While this mutation was observed in the B.1.1....  PMC8769358\n",
      "12  B.1.1.317  The A845S spike mutation was observed at the s...  PMC8769358,            0                                                  1          2\n",
      "0    B.1.351  As a result of viral evolution, the initial SA...  PPR442789\n",
      "1    B.1.1.7  As a result of viral evolution, the initial SA...  PPR442789\n",
      "2  B.1.617.2  As a result of viral evolution, the initial SA...  PPR442789\n",
      "3        P.1  As a result of viral evolution, the initial SA...  PPR442789\n",
      "4  B.1.1.529  The most recent VOC, first identified in South...  PPR442789\n",
      "5    B.1.529  Their neutralization spectrum is different and...  PPR442789\n",
      "6    B.1.617  Residues mutated in B.1.1.7, B.1.617 and B.1.3...  PPR442789,            0                                                  1           2\n",
      "0    B.1.1.7  Although neutralization of one of these antibo...  PMC8690919\n",
      "1    B.1.351  Although neutralization of one of these antibo...  PMC8690919\n",
      "2  B.1.617.2  Finally, we investigate the effects of B.1.1.7...  PMC8690919\n",
      "3        P.1  Similar histopathological findings of lung and...  PMC8690919,             0                                                  1           2\n",
      "0   B.1.617.2  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "1         P.3  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "2        C.37  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "3     B.1.351  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "4         P.2  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "5     B.1.621  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "6         P.1  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "7   B.1.617.1  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "8     B.1.525  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "9     B.1.526  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "10    B.1.1.7  Furthermore, among these lineages, WHO identif...  PMC8754802\n",
      "11  B.1.214.2  For the rest of lineages, we identified indel ...  PMC8754802\n",
      "12     B.1.36  For the rest of lineages, we identified indel ...  PMC8754802\n",
      "13      A.2.5  For the rest of lineages, we identified indel ...  PMC8754802,             0                                                  1           2\n",
      "0   B.1.617.2  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "1   B.1.617.1  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "2     B.1.617  The SARS-CoV-2 B.1.617 lineage variants, Kappa...  PMC8707919\n",
      "3        C.37  In comparison, similar modest reductions were ...  PMC8707919\n",
      "4     B.1.526  In comparison, similar modest reductions were ...  PMC8707919\n",
      "5     B.1.351  In comparison, similar modest reductions were ...  PMC8707919\n",
      "6         R.1  In comparison, similar modest reductions were ...  PMC8707919\n",
      "7         P.1  In comparison, similar modest reductions were ...  PMC8707919\n",
      "8         B.1  One of the earliest variants that is highly in...  PMC8707919\n",
      "9     B.1.429  Subsequent genomic surveillance has led to the...  PMC8707919\n",
      "10    B.1.427  The B.1.427/B.1.429 and B.1.617 lineage varian...  PMC8707919\n",
      "11        P.3  The E484K substitution found in B.1.351, B.1.5...  PMC8707919\n",
      "12    B.1.1.7  These findings extend a prior study reporting ...  PMC8707919\n",
      "13      B.1.2  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "14      C.1.2  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "15    B.1.525  Furthermore, continued surveillance is necessa...  PMC8707919\n",
      "16  B.1.617.3  Sequences corresponding to the S1/S2 region of...  PMC8707919,            0                                                  1          2\n",
      "0        P.1  Genomic surveillance of SC2 variants has been ...  PPR428671\n",
      "1      A.2.5  Here, we describe the introduction of SC2 line...  PPR428671\n",
      "2      A.2.4  In January 2021, BPHL detected the first cases...  PPR428671\n",
      "3        B.1  The A.2.5 lineage in Florida became of particu...  PPR428671\n",
      "4    B.1.427  A.2.5 also contains spike L452R, a CDC-defined...  PPR428671\n",
      "5  B.1.617.1  Allele L452R has independently emerged in seve...  PPR428671\n",
      "6    B.1.526  Allele L452R has independently emerged in seve...  PPR428671\n",
      "7  B.1.617.2  Allele L452R has independently emerged in seve...  PPR428671\n",
      "8       C.36  Allele L452R has independently emerged in seve...  PPR428671\n",
      "9    B.1.1.7  Additional mutations causing increased transmi...  PPR428671,            0                                                  1           2\n",
      "0    B.1.1.7  The alpha/B.1.1.7 SARS-CoV-2 lineage emerged i...  PMC8635972\n",
      "1    B.1.351  The incidence domination was likely due to a f...  PMC8635972\n",
      "2        P.1  The incidence domination was likely due to a f...  PMC8635972\n",
      "3    B.1.621  Some of these residue substitutions, including...  PMC8635972\n",
      "4        P.3  Some of these residue substitutions, including...  PMC8635972\n",
      "5  B.1.617.2  All the above VOC/VOI have been for the most p...  PMC8635972\n",
      "6        B.1  In order to examine whether the increased affi...  PMC8635972,            0                                                  1           2\n",
      "0  B.1.617.2  FD20 engages the RBD at an epitope distal to t...  PMC8646660\n",
      "1    B.1.1.7  FD20 engages the RBD at an epitope distal to t...  PMC8646660\n",
      "2    B.1.351  FD20 engages the RBD at an epitope distal to t...  PMC8646660\n",
      "3        P.1  FD20 engages the RBD at an epitope distal to t...  PMC8646660\n",
      "4  B.1.1.298  Y453F increases the ACE2 binding affinity, and...  PMC8646660\n",
      "5    B.1.160  A222V and S477N are described in 20A.EU1 (B.1....  PMC8646660\n",
      "6    B.1.177  A222V and S477N are described in 20A.EU1 (B.1....  PMC8646660\n",
      "7    B.1.141  N439K is in B.1.141, B.1.258, and in mink stra...  PMC8646660\n",
      "8    B.1.258  N439K is in B.1.141, B.1.258, and in mink stra...  PMC8646660,             0                                                  1           2\n",
      "0        C.37  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "1     B.1.351  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "2     B.1.1.7  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "3     B.1.621  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "4   B.1.617.2  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "5         P.1  Four SARS-CoV-2 variants of concern including ...  PMC8602852\n",
      "6      C.36.3  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "7   B.1.617.1  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "8   B.1.1.519  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "9   B.1.214.2  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "10    B.1.525  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "11    B.1.619  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "12    B.1.526  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "13  B.1.1.318  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "14        R.1  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "15  B.1.466.2  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "16      C.1.2  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "17    B.1.620  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "18  B.1.1.523  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "19    B.1.427  Currently designated VUMs include B.1.427/B.1....  PMC8602852\n",
      "20        P.2  Some former VOIs but no longer designated as V...  PMC8602852\n",
      "21        P.3  Some former VOIs but no longer designated as V...  PMC8602852\n",
      "22   B.1.1.28  P.3 as well as P.1 belong to B.1.1.28 lineage ...  PMC8602852\n",
      "23    B.1.617  At the same time, B.1.525 was resistant to neu...  PMC8602852\n",
      "24  B.1.617.3  With the spread of the local epidemic in India...  PMC8602852,            0                                                  1           2\n",
      "0  B.1.617.2  The Delta (B.1.617.2) variant was first identi...  PMC8598544\n",
      "1       C.37  The Lambda variant belongs to the C.37 lineage...  PMC8598544\n",
      "2    B.1.621  On August 30, 2021, the WHO announced the Mu v...  PMC8598544,            0                                                  1           2\n",
      "0    B.1.351  The new variants are accumulating mutations in...  PMC8535385\n",
      "1    B.1.1.7  The new variants are accumulating mutations in...  PMC8535385\n",
      "2  B.1.617.2  The new variants are accumulating mutations in...  PMC8535385\n",
      "3        P.1  The new variants are accumulating mutations in...  PMC8535385,          0                                                  1           2\n",
      "0  B.1.117  At the time of the first clinical trials the “...  PMC8501352, Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],              0                                                  1           2\n",
      "0      B.1.427  The massive emergence of L452R variants was fi...  PMC8525575\n",
      "1      B.1.232  By PCR amplification and Sanger sequencing of ...  PMC8525575\n",
      "2     B.1.1.28  Because of the important functional and antige...  PMC8525575\n",
      "3      B.1.351  Because of the important functional and antige...  PMC8525575\n",
      "4      B.1.1.7  Because of the important functional and antige...  PMC8525575\n",
      "5      A.2.5.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "6    B.1.617.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "7         A.21  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "8         C.37  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "9    B.1.617.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "10         P.4  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "11      C.36.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "12   B.1.617.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "13     B.1.429  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "14      C.36.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "15      B.1.36  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "16        C.16  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "17        A.27  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "18   B.1.526.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "19     B.1.466  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "20   B.1.1.263  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "21  B.1.177.83  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "22   B.1.1.459  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "23     A.2.5.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575,           0                                                  1          2\n",
      "0   B.1.351  In late 2020, three SARS-CoV-2 variants of con...  PPR380399\n",
      "1   B.1.1.7  In late 2020, three SARS-CoV-2 variants of con...  PPR380399\n",
      "2       P.1  In late 2020, three SARS-CoV-2 variants of con...  PPR380399\n",
      "3   B.1.526  In early 2021, several new VOC appeared includ...  PPR380399\n",
      "4   B.1.617  In early 2021, several new VOC appeared includ...  PPR380399\n",
      "5   B.1.427  In early 2021, several new VOC appeared includ...  PPR380399\n",
      "6       D.2  The D.2 variant in Australia was defined by an...  PPR380399\n",
      "7   B.1.160  In contrast, B.1.160, which carries S S477N so...  PPR380399\n",
      "8   B.1.429  The substitution D260Y, present in B.1.427 and...  PPR380399\n",
      "9   B.1.177  Several lineages show introgression from other...  PPR380399\n",
      "10      B.1  c. Population-level analysis of new mutations ...  PPR380399, Empty DataFrame\n",
      "Columns: []\n",
      "Index: [], Empty DataFrame\n",
      "Columns: []\n",
      "Index: [],       0                                                  1           2\n",
      "0   V.2  Variants with minor allele frequency less than...  PMC7705438\n",
      "1  C.14  We identified one previously reported (c.2729 ...  PMC7705438\n",
      "2   V.1  All variants were reclassified according to AC...  PMC7705438\n",
      "3  C.38  Results: We found 8 mutations in this patient ...  PMC7705438, Empty DataFrame\n",
      "Columns: []\n",
      "Index: []][            0                                                  1           2\n",
      "0         P.1  The most common genotype was 417variant/484K/5...  PMC9145602\n",
      "1         P.2  No mutations (K417 only) were found in 64/198 ...  PMC9145602\n",
      "2     B.1.1.7  Seven samples (3.5%) tested positive for 452R ...  PMC9145602\n",
      "3   B.1.617.2  Patient consent was waived under approval of I...  PMC9145602\n",
      "4   B.1.617.1  Patient consent was waived under approval of I...  PMC9145602\n",
      "5        C.37  Patient consent was waived under approval of I...  PMC9145602\n",
      "6     B.1.351  The most common genotype identified was 417var...  PMC9145602\n",
      "7       P.1.2  Samples bearing genotype 417variant/484K/501Y ...  PMC9145602\n",
      "8    B.1.1.28  The genotype of only K417 was found in samples...  PMC9145602\n",
      "9     B.1.499  The genotype of only K417 was found in samples...  PMC9145602\n",
      "10  B.1.1.277  The genotype of only K417 was found in samples...  PMC9145602\n",
      "11   B.1.1.33  The genotype of only K417 was found in samples...  PMC9145602\n",
      "12    A.2.5.2  A lineage from 2020 that was still circulating...  PMC9145602\n",
      "13  B.1.1.529  While this mutation is distinct from T478K fou...  PMC9145602,            0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647,              0                                                  1           2\n",
      "0      B.1.427  The massive emergence of L452R variants was fi...  PMC8525575\n",
      "1      B.1.232  By PCR amplification and Sanger sequencing of ...  PMC8525575\n",
      "2     B.1.1.28  Because of the important functional and antige...  PMC8525575\n",
      "3      B.1.1.7  Because of the important functional and antige...  PMC8525575\n",
      "4      B.1.351  Because of the important functional and antige...  PMC8525575\n",
      "5    B.1.617.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "6       B.1.36  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "7   B.1.177.83  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "8    B.1.1.459  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "9          P.4  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "10   B.1.617.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "11   B.1.1.263  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "12      C.36.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "13        A.21  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "14   B.1.526.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "15      C.36.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "16     A.2.5.1  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "17     B.1.429  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "18   B.1.617.3  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "19        C.37  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "20        C.16  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "21     A.2.5.2  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "22     B.1.466  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575\n",
      "23        A.27  A.2.5, A.2.5.1, A.2.5.2, A.21, A.27, B.1.1.263...  PMC8525575]\n",
      "done with A.2.5.2, time duration --- seconds --- 27.01111102104187 \n",
      "\n",
      "start A.2.5.3\n",
      "{\"ver\n",
      "[           0                                                  1           2\n",
      "0    B.1.1.7  This strain clade 201/501Y.V1, Pango lineage B...  PMC9088647\n",
      "1    B.1.351  26\\n There are three mutations of significant ...  PMC9088647\n",
      "2        P.1  56\\n Safety and immunogenicity study of a SARS...  PMC9088647\n",
      "3        P.2  Gamma variant is particularly found from the o...  PMC9088647\n",
      "4    B.1.429  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "5    B.1.427  These resembling variants, B.1.427 and B.1.429...  PMC9088647\n",
      "6    B.1.617  B.1.617.2 (Delta variant) is a variant of line...  PMC9088647\n",
      "7  B.1.617.1  83\\n The B.1.617.1 strain is 6.8‐fold more res...  PMC9088647\n",
      "8      C.1.2  92\\n Researchers discovered that this new C.1....  PMC9088647\n",
      "9  B.1.1.529  K417N, N440K, G446S, S477N, T478K, E484A, Q493...  PMC9088647]\n",
      "done with A.2.5.3, time duration --- seconds --- 9.311837911605835 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "for i in l_10[6:9]:\n",
    "    iter_lineage(i)\n",
    "end = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time: 67.0862021446228\n"
     ]
    }
   ],
   "source": [
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('A.2.5.1',)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b = db.from_sequence(l_10[6:9])\n",
    "b.take(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[None, None, None]\n",
      "Total time: 27.45674777030945\n"
     ]
    }
   ],
   "source": [
    "b = db.from_sequence(l_10[6:9])\n",
    "start = time.time()\n",
    "result = b.map(lambda l: iter_lineage(l))\n",
    "print(result.compute())\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bag lineage and run whole procude --> runtime error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start A.2.5.1\n",
      "{\"ver\n",
      "done with A.2.5.1, time duration --- seconds --- 24.798223733901978 \n",
      "\n",
      "start A.2.4\n",
      "{\"ver\n",
      "done with A.2.4, time duration --- seconds --- 128.26828908920288 \n",
      "\n",
      "start A.2.5\n",
      "{\"ver\n",
      "done with A.2.5, time duration --- seconds --- 582.8988699913025 \n",
      "\n",
      "start A.2.3\n",
      "{\"ver\n",
      "done with A.2.3, time duration --- seconds --- 58.731218099594116 \n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_73404/3436830193.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0miter_lineage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Total time:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m         \"\"\"\n\u001b[0;32m--> 290\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mpostcomputes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dask_postcompute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 573\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    574\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)\u001b[0m\n\u001b[1;32m   2992\u001b[0m                     \u001b[0mshould_rejoin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2993\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2994\u001b[0;31m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2995\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2996\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfutures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mgather\u001b[0;34m(self, futures, errors, direct, asynchronous)\u001b[0m\n\u001b[1;32m   2150\u001b[0m                 \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2151\u001b[0m                 \u001b[0mlocal_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlocal_worker\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2152\u001b[0;31m                 \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2153\u001b[0m             )\n\u001b[1;32m   2154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    308\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    309\u001b[0m             return sync(\n\u001b[0;32m--> 310\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    311\u001b[0m             )\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    370\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_set\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 372\u001b[0;31m             \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(timeout)\u001b[0m\n\u001b[1;32m    359\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0mloop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcancel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    550\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 552\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    553\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    554\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    298\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "b = db.from_sequence(l_10)\n",
    "start = time.time()\n",
    "result = b.map(lambda l: iter_lineage(l))\n",
    "result.compute()\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Break down and parallel on id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parellel_iter1(l_): \n",
    "    \n",
    "    #for l_ in lineage: This is redundant in mapping\n",
    "     \n",
    "    start = time.time()\n",
    "    print(f'start {l_}')\n",
    "    #get ids\n",
    "    ids = get_ids(l_)\n",
    "\n",
    "    try:\n",
    "        # if lineage processed before, run on new ids\n",
    "        path = Path(f'{l_}_df.csv')\n",
    "        if path.is_file():\n",
    "            df = pd.read_csv(path)\n",
    "            id_ran = set(df.ID)\n",
    "            id_to_run = set(ids) - id_ran\n",
    "            if id_to_run: \n",
    "                bag_id = db.from_sequence(id_to_run)\n",
    "                dds = bag_id.map(lambda i: extract_full_parallel2(i))\n",
    "                df_new = dds.take(1)[0]\n",
    "                ddf = pd.concat([df, df_new], axis = 0)\n",
    "\n",
    "        else: # otherwise run on all ids\n",
    "            id_to_run = ids\n",
    "            bag_id = db.from_sequence(id_to_run)\n",
    "            dds = bag_id.map(lambda i: extract_full_parallel2(i)) \n",
    "            ddf = dds.take(1)[0]\n",
    "\n",
    "        ddf.to_dataframe().to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                    header = ['lineage', 'string', 'ID'])\n",
    "        end = time.time()\n",
    "        print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "    except:\n",
    "        id_to_run = ids\n",
    "        bag_id = db.from_sequence(id_to_run)\n",
    "        dds = bag_id.map(lambda i: extract_full_parallel2(i))\n",
    "        ddf = dds.take(1)[0]\n",
    "        ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                    header = ['lineage', 'string', 'ID'])\n",
    "        end = time.time()\n",
    "        print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_full(ids):\n",
    "    full_regrex = []\n",
    "    \n",
    "    if not ids:  \n",
    "        return None\n",
    "    for i in ids:\n",
    "        try:\n",
    "            body_text = download_article(i) # get body text\n",
    "            record = get_full_lineage(body_text) # extract lineages in text\n",
    "            [x.append(i) for x in record] # attach article id to lineage record\n",
    "            full_regrex.append(pd.DataFrame(record))\n",
    "        except urllib.error.HTTPError as exc:\n",
    "            time.sleep(5) # wait 10 seconds and then make http request again\n",
    "            continue\n",
    "    df_fulltext = pd.concat(full_regrex)\n",
    "    df_fulltext.columns = ['lineage', 'string', 'ID']\n",
    "    return df_fulltext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# no looping\n",
    "def extract_full_parallel(ids):\n",
    "    #ids = dic['ids']\n",
    "    full_regrex = []\n",
    "    \n",
    "    \"\"\"if not ids:  \n",
    "        return None\n",
    "    for i in ids:\"\"\"\n",
    "    \n",
    "    try:\n",
    "        body_text = download_article(ids) # get body text\n",
    "        record = get_full_lineage(body_text) # extract lineages in text\n",
    "        [x.append(ids) for x in record] # attach article id to lineage record\n",
    "        full_regrex.append(pd.DataFrame(record))\n",
    "    except urllib.error.HTTPError as exc:\n",
    "        time.sleep(5) # wait 5 seconds and then make http request again\n",
    "        #continue\n",
    "    \n",
    "    \n",
    "    df_fulltext = pd.DataFrame(full_regrex)\n",
    "    print(df_fulltext)\n",
    "    df_fulltext.columns = ['lineage', 'string', 'ID']\n",
    "    return df_fulltext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# looping over ids\n",
    "def extract_full_parallel2(ids):\n",
    "    #ids = dic['ids']\n",
    "    full_regrex = []\n",
    "    \"\"\"\n",
    "    if not ids:  \n",
    "        return None\n",
    "    \"\"\"\n",
    "    for i in ids:\n",
    "        try:\n",
    "            body_text = download_article(i) # get body text\n",
    "            record = get_full_lineage(body_text) # extract lineages in text\n",
    "            [x.append(i) for x in record] # attach article id to lineage record\n",
    "            full_regrex.append(pd.DataFrame(record))\n",
    "        except urllib.error.HTTPError as exc:\n",
    "            time.sleep(5) # wait 10 seconds and then make http request again\n",
    "            continue\n",
    "    df_fulltext = pd.concat(full_regrex)\n",
    "    df_fulltext.columns = ['lineage', 'string', 'ID']\n",
    "    return df_fulltext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### BAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query(lineage):\n",
    "    # get articles ids for specific lineage:\n",
    "    url = requests.get(f'https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=(%22{lineage}%22%20AND%20(%22SARS-CoV-2%22%20OR%20%22COVID-19%22)%20AND%20(%22lineage%22%20OR%20%22lineages%22%20OR%20%22strain%22%20OR%20%22strains%22%20OR%20%22variants%22%20OR%20%22variants%22))%20AND%20(FIRST_PDATE:%5b2020-01-01%20)%20AND%20HAS_FT:y%20AND%20%20sort_date:y&resultType=idlist&pageSize=1000&format=json&cursorMark=*')\n",
    "    text = url.text\n",
    "    print(text[:10])\n",
    "    results = json.loads(text)['resultList']['result']\n",
    "    ids = list(map(lambda x: x['fullTextIdList']['fullTextId'][0], results))\n",
    "    return {'lineage': lineage, 'ids': ids}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineages[7:8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## try if work on lineage A.1 & A2.5.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parallel_iter(lineages): \n",
    "    \n",
    "    for l_ in lineages: #This is redundant in mapping\n",
    "        start = time.time()\n",
    "        print(f'start {l_}')\n",
    "        #get ids\n",
    "        ids = get_ids(l_)\n",
    "\n",
    "        try:\n",
    "            # if lineage processed before, run on new ids\n",
    "            path = Path(f'{l_}_df.csv')\n",
    "            if path.is_file():\n",
    "                print('exist')\n",
    "                df = pd.read_csv(path)\n",
    "                id_to_run = set(ids) - set(df.ID) # ids need to be processed\n",
    "                print(f'new id {id_to_run}')\n",
    "               \n",
    "                # partition ids and run on parallel\n",
    "                N = 10\n",
    "                if len(id_to_run) < 12:\n",
    "                    N = 1\n",
    "                id_bag = db.from_sequence(ids, npartitions = N)\n",
    "                id_bag.map(lambda i: extract_full(i))\n",
    "                df_new = extract_full(id_to_run) # merge with original df\n",
    "                print(df_new)\n",
    "                ddf = pd.concat([df, df_new], axis = 0)\n",
    "                print('merged')\n",
    "                #if id_ran != set(ids): \n",
    "                    #print('new id to run')\n",
    "                    #dds = result.filter(lambda x: x not in list(id_ran)).map(lambda i: extract_full_parallel2(i))\n",
    "                    #df_new = dds.take(1)[0]\n",
    "\n",
    "                #else: return # if no updates do nothing\n",
    "                ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                            header = ['lineage', 'string', 'ID'])\n",
    "                end = time.time()\n",
    "                print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "            else: # otherwise run on all ids\n",
    "                raise Exception('run on all')\n",
    "        except:\n",
    "            #result = db.from_sequence(ids)\n",
    "            dds = extract_full(ids)\n",
    "            #ddf = dds.take(1)[0]\n",
    "            ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
    "                                        header = ['lineage', 'string', 'ID'])\n",
    "            end = time.time()\n",
    "            print(f'done with {l_}, time duration --- seconds --- {end - start} \\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('PMC9205280', 'PMC8251645')"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = {'PMC9205280', 'PMC8251645', 'PMC7650373', 'PMC7741497'}\n",
    "b = db.from_sequence(ids,npartitions=2)\n",
    "b.take(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n",
      "start A.2.2\n",
      "exist\n",
      "new id {'PMC9205280', 'PMC8251645', 'PMC7650373', 'PMC7741497'}\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/506483837.py\u001b[0m in \u001b[0;36mparallel_iter\u001b[0;34m(l_)\u001b[0m\n\u001b[1;32m     21\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'new id {id_to_run}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m             \u001b[0mdf_new\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_full\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mid_to_run\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# merge with original df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_new\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/2230385743.py\u001b[0m in \u001b[0;36mextract_full\u001b[0;34m(ids)\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0mdf_fulltext\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfull_regrex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m     \u001b[0mdf_fulltext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'lineage'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'string'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdf_fulltext\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   5499\u001b[0m             \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5500\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5501\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/pandas/_libs/properties.pyx\u001b[0m in \u001b[0;36mpandas._libs.properties.AxisProperty.__set__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_set_axis\u001b[0;34m(self, axis, labels)\u001b[0m\n\u001b[1;32m    765\u001b[0m         \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 766\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    767\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mset_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m    215\u001b[0m         \u001b[0;31m# Caller is responsible for ensuring we have an Index object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 216\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    217\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/pandas/core/internals/base.py\u001b[0m in \u001b[0;36m_validate_set_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m     57\u001b[0m             raise ValueError(\n\u001b[0;32m---> 58\u001b[0;31m                 \u001b[0;34mf\"Length mismatch: Expected axis has {old_len} elements, new \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m                 \u001b[0;34mf\"values have {new_len} elements\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Length mismatch: Expected axis has 0 elements, new values have 3 elements",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/3752159325.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mparallel_iter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'A.2.2'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/506483837.py\u001b[0m in \u001b[0;36mparallel_iter\u001b[0;34m(l_)\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m         \u001b[0;31m#result = db.from_sequence(ids)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0mdds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_full\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m         \u001b[0;31m#ddf = dds.take(1)[0]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         ddf.to_csv(f'{l_}_df.csv',index=False, \\\n",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/2230385743.py\u001b[0m in \u001b[0;36mextract_full\u001b[0;34m(ids)\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mids\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m             \u001b[0mbody_text\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdownload_article\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# get body text\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m             \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_full_lineage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbody_text\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# extract lineages in text\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m             \u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrecord\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# attach article id to lineage record\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/3816319870.py\u001b[0m in \u001b[0;36mdownload_article\u001b[0;34m(article_id)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mdownload_article\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marticle_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0murl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf'https://www.ebi.ac.uk/europepmc/webservices/rest/{article_id}/fullTextXML'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mxmldoc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murlopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;31m# get full text\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/xml/etree/ElementTree.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(source, parser)\u001b[0m\n\u001b[1;32m   1195\u001b[0m     \"\"\"\n\u001b[1;32m   1196\u001b[0m     \u001b[0mtree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mElementTree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1197\u001b[0;31m     \u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1198\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mtree\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1199\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/xml/etree/ElementTree.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(self, source, parser)\u001b[0m\n\u001b[1;32m    596\u001b[0m                     \u001b[0;31m# It can be used to parse the whole source without feeding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    597\u001b[0m                     \u001b[0;31m# it with chunks.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 598\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_root\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parse_whole\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    599\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_root\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    600\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/http/client.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    463\u001b[0m             \u001b[0;31m# Amount is given, implement using readinto\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    464\u001b[0m             \u001b[0mb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbytearray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mamt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 465\u001b[0;31m             \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadinto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    466\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mmemoryview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtobytes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    467\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/http/client.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    497\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    498\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchunked\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 499\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_readinto_chunked\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    500\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    501\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/http/client.py\u001b[0m in \u001b[0;36m_readinto_chunked\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    602\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    603\u001b[0m                 \u001b[0mtemp_mvb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmvb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mchunk_left\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 604\u001b[0;31m                 \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_safe_readinto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp_mvb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    605\u001b[0m                 \u001b[0mmvb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmvb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    606\u001b[0m                 \u001b[0mtotal_bytes\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/http/client.py\u001b[0m in \u001b[0;36m_safe_readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    642\u001b[0m                 \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadinto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp_mvb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    643\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 644\u001b[0;31m                 \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadinto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmvb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    645\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    646\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mIncompleteRead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbytes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmvb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mtotal_bytes\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    587\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    588\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 589\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    590\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    591\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/ssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1069\u001b[0m                   \u001b[0;34m\"non-zero flags not allowed in calls to recv_into() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1070\u001b[0m                   self.__class__)\n\u001b[0;32m-> 1071\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1072\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1073\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/ssl.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m    927\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    928\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbuffer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 929\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    930\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    931\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "parallel_iter('A.2.2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start A.2.5.2\n",
      "{\"ver\n",
      "exist\n",
      "new id {'PMC9145602'}\n",
      "done merge\n",
      "done with A.2.5.2, time duration --- seconds --- 19.290800094604492 \n",
      "\n",
      "Total time: 19.291192054748535\n"
     ]
    }
   ],
   "source": [
    "s = pd.read_csv('A.2.5.2_df.csv')\n",
    "s[15:].to_csv('A.2.5.2_df.csv')\n",
    "start = time.time()\n",
    "iter_lineage('A.2.5.2')\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exist\n",
      "{'PMC9088647', 'PMC8525575'}\n",
      "['PMC9145602', 'PMC9088647', 'PMC8525575']\n",
      "here\n",
      "found\n",
      "Total time: 15.104039907455444\n"
     ]
    }
   ],
   "source": [
    "# break down iter code for debugging\n",
    "l_ = 'A.2.5.2'\n",
    "start = time.time()\n",
    "b = db.from_sequence([l_])\n",
    "result = b.map(lambda record: get_ids(record))\n",
    "path = Path(f'{l_}_df.csv')\n",
    "if path.is_file():\n",
    "    print('exist')\n",
    "    df = pd.read_csv(path)\n",
    "    ids = result.take(1)[0]\n",
    "    \n",
    "    id_ran = set(df.ID)\n",
    "    print(id_ran)\n",
    "    print(ids)\n",
    "    if id_ran != set(ids): \n",
    "        print('here')\n",
    "        dds = result.filter(lambda x: x not in list(id_ran)).map(lambda i: extract_full_parallel2(i))\n",
    "        print('found')\n",
    "        df_new = dds.take(1)[0]\n",
    "        ddf = pd.concat([df, df_new], axis = 0)\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start A.2.2\n",
      "{\"ver\n",
      "exist\n",
      "new id {'PMC8920968', 'PMC7650373', 'PMC9205280', 'PMC9012253', 'PMC9230982', 'PMC9088647', 'PMC8251645', 'PMC7741497'}\n",
      "done merge\n",
      "done with A.2.2, time duration --- seconds --- 212.11726903915405 \n",
      "\n",
      "start A.2.3\n",
      "{\"ver\n",
      "exist\n",
      "new id {'PPR417362', 'PMC8662802', 'PMC8390340', 'PMC9088647', 'PMC7350918', 'PMC9174147'}\n",
      "done merge\n",
      "done with A.2.3, time duration --- seconds --- 79.7303831577301 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "for i in l_10[2:4]:\n",
    "    iter_lineage(i)\n",
    "end = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "end-start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = pd.read_csv('A.2.2_df.csv')\n",
    "s[30:].to_csv('A.2.2_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = pd.read_csv('A.2.3_df.csv')\n",
    "s[30:].to_csv('A.2.3_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time: 3.3931920528411865\n"
     ]
    }
   ],
   "source": [
    "# parallel\n",
    "b = db.from_sequence(lineages[6:9])\n",
    "start = time.time()\n",
    "result = b.map(lambda record: parallel_iter(record))\n",
    "result.compute()\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "distributed.worker - WARNING - Compute Failed\n",
      "Function:  execute_task\n",
      "args:      ((<function safe_take at 0x7fa6d210ad40>, 1, (<function map_chunk at 0x7fa6d210a560>, <function parallel_iter.<locals>.<lambda> at 0x7fa6d6c504d0>, [['PMC9174147']], None, {}), True))\n",
      "kwargs:    {}\n",
      "Exception: 'NameError(\"name \\'i\\' is not defined\")'\n",
      "\n",
      "distributed.worker - WARNING - Compute Failed\n",
      "Function:  execute_task\n",
      "args:      ((<function safe_take at 0x7fa4959d7680>, 1, (<function map_chunk at 0x7fa4959d6e60>, <function parallel_iter.<locals>.<lambda> at 0x7fa49a09c050>, [['PMC9230982']], None, {}), True))\n",
      "kwargs:    {}\n",
      "Exception: 'NameError(\"name \\'i\\' is not defined\")'\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_97801/1281572401.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mrecord\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mparallel_iter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Total time:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m         \"\"\"\n\u001b[0;32m--> 290\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mpostcomputes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dask_postcompute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 573\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    574\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)\u001b[0m\n\u001b[1;32m   2992\u001b[0m                     \u001b[0mshould_rejoin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2993\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2994\u001b[0;31m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2995\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2996\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfutures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mgather\u001b[0;34m(self, futures, errors, direct, asynchronous)\u001b[0m\n\u001b[1;32m   2150\u001b[0m                 \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2151\u001b[0m                 \u001b[0mlocal_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlocal_worker\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2152\u001b[0;31m                 \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2153\u001b[0m             )\n\u001b[1;32m   2154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    308\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    309\u001b[0m             return sync(\n\u001b[0;32m--> 310\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    311\u001b[0m             )\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    370\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_set\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 372\u001b[0;31m             \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(timeout)\u001b[0m\n\u001b[1;32m    359\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0mloop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcancel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    550\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 552\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    553\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    554\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    298\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "b = db.from_sequence(lineages[2:4])\n",
    "start = time.time()\n",
    "result = b.map(lambda record: parallel_iter(record))\n",
    "result.compute()\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with new id added: 179\n",
    "# nothing changed: \n",
    "# start new: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## One example of break down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = pd.read_csv('A.2.5.1_df.csv')\n",
    "s[:-30].to_csv('A.2.5.1_df.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start A.2.5.1\n",
      "exist\n",
      "new id to run\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "distributed.worker - WARNING - Compute Failed\n",
      "Function:  execute_task\n",
      "args:      ((<function safe_take at 0x7fac6cb895f0>, 1, (<function map_chunk at 0x7fac6cb88dd0>, <function parallel_iter.<locals>.<lambda> at 0x7fac6c376a70>, [(<class 'filter'>, <function parallel_iter.<locals>.<lambda> at 0x7fac6c22fd40>, (<function map_chunk at 0x7fac6cb88dd0>, <function parallel_iter.<locals>.<lambda> at 0x7fac6c22fcb0>, [['A.2.5.1']], None, {}))], None, {}), True))\n",
      "kwargs:    {}\n",
      "Exception: 'TypeError(\"unhashable type: \\'list\\'\")'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with A.2.5.1, time duration --- seconds --- 16.260791063308716 \n",
      "\n",
      "Total time: 16.26163601875305\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "parallel_iter('A.2.5.1')\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A.2.5.1' 'A.2.5.2' 'A.2.5.3']\n",
      "('A.2.2',)\n",
      "get id\n",
      "(['PMC9230982', 'PMC9205280', 'PMC9012253', 'PMC9088647', 'PMC8920968', 'PMC8883582', 'PPR453676', 'PPR439656', 'PPR439367', 'PMC8688474', 'PPR425039', 'PMC8561373', 'PMC8251645', 'PPR361607', 'PMC7705438', 'PMC7650373', 'PMC7741497'],)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/fb/g1bkxzgx7l3f2b_yv97h4h_40000gn/T/ipykernel_13438/1523319328.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mdds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mextract_full_parallel2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Total time:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m         \"\"\"\n\u001b[0;32m--> 290\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    571\u001b[0m         \u001b[0mpostcomputes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dask_postcompute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 573\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    574\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)\u001b[0m\n\u001b[1;32m   2992\u001b[0m                     \u001b[0mshould_rejoin\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2993\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2994\u001b[0;31m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpacked\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2995\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2996\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfutures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mgather\u001b[0;34m(self, futures, errors, direct, asynchronous)\u001b[0m\n\u001b[1;32m   2150\u001b[0m                 \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2151\u001b[0m                 \u001b[0mlocal_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlocal_worker\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2152\u001b[0;31m                 \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2153\u001b[0m             )\n\u001b[1;32m   2154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    308\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    309\u001b[0m             return sync(\n\u001b[0;32m--> 310\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    311\u001b[0m             )\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    370\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_set\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 372\u001b[0;31m             \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(timeout)\u001b[0m\n\u001b[1;32m    359\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0mloop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcancel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    550\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 552\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    553\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    554\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/envs/covid-19-community/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    298\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 300\u001b[0;31m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    301\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m                     \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import dask.bag as db\n",
    "b = db.from_sequence(lineages[2:4])\n",
    "print(lineages[6:9])\n",
    "print(b.take(2))\n",
    "start = time.time()\n",
    "result = b.map(lambda record: get_ids(record))\n",
    "print('get id')\n",
    "print(result.take(1))\n",
    "dds = result.map(lambda x: extract_full_parallel2(x))\n",
    "print(dds.compute())\n",
    "end = time.time()\n",
    "print('Total time:', end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.take(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = db.from_sequence([lineages[7:8]])\n",
    "print(lineages[7:8])\n",
    "start = time.time()\n",
    "result = b.map(lambda record: query(record))\n",
    "print('get id')\n",
    "ddf = result.map(lambda x: extract_full_parallel(x))\n",
    "ddf.compute()\n",
    "end = time.time()\n",
    "print(end - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get dataframe\n",
    "ddf.take(1)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download articles in XML and return body paragraph\n",
    "def process(article_id):\n",
    "    url = f'https://www.ebi.ac.uk/europepmc/webservices/rest/{article_id}/fullTextXML'\n",
    "    xmldoc = parse(urlopen(url))\n",
    "    \n",
    "    # get full text\n",
    "    root = xmldoc.getroot()\n",
    "    text = root.findall('.//p')\n",
    "\n",
    "    # put body paragraphs together\n",
    "    ptext = \"\"\n",
    "    for p in text:\n",
    "        ptext += ''.join([x for x in p.itertext()]) + '.\\n' + '\\n'\n",
    "    \n",
    "    # tokenize texts into sentences\n",
    "    p_sentence = nltk.tokenize.sent_tokenize(ptext)\n",
    "    \n",
    "    # record lineages\n",
    "    linset = set()\n",
    "    pair = []\n",
    "    for s in p_sentence:\n",
    "        s1 = re.subn('[()/,]', ' ', s)[0] # remove special chars\n",
    "        lin = set(pattern1.findall(s1) + pattern2.findall(s1) + pattern3.findall(s1) + pattern4.findall(s1))\n",
    "\n",
    "        if lin: \n",
    "            for l in lin:\n",
    "                # valid lineage and not recorded\n",
    "                l = l.strip()\n",
    "                l = l.capitalize()\n",
    "                if (l in lineages) and (l not in linset): \n",
    "                    linset.add(l)\n",
    "                    pair.append([l, s])\n",
    "                else: continue\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get lineage for full texts\n",
    "def get_full_lineage(ptext):\n",
    "    # tokenize texts into sentences\n",
    "    p_sentence = nltk.tokenize.sent_tokenize(ptext)\n",
    "    \n",
    "    # record lineages\n",
    "    linset = set()\n",
    "    pair = []\n",
    "    for s in p_sentence:\n",
    "        s1 = re.subn('[()/,]', ' ', s)[0] # remove special chars\n",
    "        lin = set(pattern1.findall(s1) + pattern2.findall(s1) + pattern3.findall(s1) + pattern4.findall(s1))\n",
    "\n",
    "        if lin: \n",
    "            for l in lin:\n",
    "                # valid lineage and not recorded\n",
    "                l = l.strip()\n",
    "                l = l.capitalize()\n",
    "                if (l in lineages) and (l not in linset): \n",
    "                    linset.add(l)\n",
    "                    pair.append([l, s])\n",
    "                else: continue\n",
    "    return pair\n",
    "    \n",
    "    \"\"\"\n",
    "    ptext = re.subn('[()/,]', ' ', ptext)[0] # remove special chars\n",
    "    lin = pattern1.findall(ptext) + pattern2.findall(ptext) + pattern3.findall(ptext)\n",
    "    lin_set = set(lin)\n",
    "    \n",
    "    record = []\n",
    "    if lin_set:\n",
    "        for l in lin_set:\n",
    "            \n",
    "            sen = re.search(r\"\\.?([^\\.]*{}[^\\.]*)\".format(l), ptext).group()\n",
    "            record.append([l, sen])\n",
    "    \"\"\"\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wrap up function take lineage ids as input and output dataframe\n",
    "def extract_full(ids):\n",
    "    full_regrex = []\n",
    "    \n",
    "    if not ids:  \n",
    "        return None\n",
    "    for i in ids:\n",
    "        try:\n",
    "            body_text = download_article(i) # get body text\n",
    "            record = get_full_lineage(body_text) # extract lineages in text\n",
    "            [x.append(i) for x in record] # attach article id to lineage record\n",
    "            full_regrex.append(pd.DataFrame(record))\n",
    "        except urllib.error.HTTPError as exc:\n",
    "            time.sleep(10) # wait 10 seconds and then make http request again\n",
    "            continue\n",
    "    df_fulltext = pd.concat(full_regrex)\n",
    "    df_fulltext.columns = ['lineage', 'string', 'ID']\n",
    "    return df_fulltext\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.map(lambda x: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
